<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Magnus Dehli Vigeland" />

<meta name="date" content="2022-11-03" />

<title>Introduction to the pedtools package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>$(document).ready(function(){
    if (typeof $('[data-toggle="tooltip"]').tooltip === 'function') {
        $('[data-toggle="tooltip"]').tooltip();
    }
    if ($('[data-toggle="popover"]').popover === 'function') {
        $('[data-toggle="popover"]').popover();
    }
});
</script>
<style type="text/css">
.lightable-minimal {
border-collapse: separate;
border-spacing: 16px 1px;
width: 100%;
margin-bottom: 10px;
}
.lightable-minimal td {
margin-left: 5px;
margin-right: 5px;
}
.lightable-minimal th {
margin-left: 5px;
margin-right: 5px;
}
.lightable-minimal thead tr:last-child th {
border-bottom: 2px solid #00000050;
empty-cells: hide;
}
.lightable-minimal tbody tr:first-child td {
padding-top: 0.5em;
}
.lightable-minimal.lightable-hover tbody tr:hover {
background-color: #f5f5f5;
}
.lightable-minimal.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-classic {
border-top: 0.16em solid #111111;
border-bottom: 0.16em solid #111111;
width: 100%;
margin-bottom: 10px;
margin: 10px 5px;
}
.lightable-classic tfoot tr td {
border: 0;
}
.lightable-classic tfoot tr:first-child td {
border-top: 0.14em solid #111111;
}
.lightable-classic caption {
color: #222222;
}
.lightable-classic td {
padding-left: 5px;
padding-right: 5px;
color: #222222;
}
.lightable-classic th {
padding-left: 5px;
padding-right: 5px;
font-weight: normal;
color: #222222;
}
.lightable-classic thead tr:last-child th {
border-bottom: 0.10em solid #111111;
}
.lightable-classic.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-classic.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-classic-2 {
border-top: 3px double #111111;
border-bottom: 3px double #111111;
width: 100%;
margin-bottom: 10px;
}
.lightable-classic-2 tfoot tr td {
border: 0;
}
.lightable-classic-2 tfoot tr:first-child td {
border-top: 3px double #111111;
}
.lightable-classic-2 caption {
color: #222222;
}
.lightable-classic-2 td {
padding-left: 5px;
padding-right: 5px;
color: #222222;
}
.lightable-classic-2 th {
padding-left: 5px;
padding-right: 5px;
font-weight: normal;
color: #222222;
}
.lightable-classic-2 tbody tr:last-child td {
border-bottom: 3px double #111111;
}
.lightable-classic-2 thead tr:last-child th {
border-bottom: 1px solid #111111;
}
.lightable-classic-2.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-classic-2.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-material {
min-width: 100%;
white-space: nowrap;
table-layout: fixed;
font-family: Roboto, sans-serif;
border: 1px solid #EEE;
border-collapse: collapse;
margin-bottom: 10px;
}
.lightable-material tfoot tr td {
border: 0;
}
.lightable-material tfoot tr:first-child td {
border-top: 1px solid #EEE;
}
.lightable-material th {
height: 56px;
padding-left: 16px;
padding-right: 16px;
}
.lightable-material td {
height: 52px;
padding-left: 16px;
padding-right: 16px;
border-top: 1px solid #eeeeee;
}
.lightable-material.lightable-hover tbody tr:hover {
background-color: #f5f5f5;
}
.lightable-material.lightable-striped tbody tr:nth-child(even) {
background-color: #f5f5f5;
}
.lightable-material.lightable-striped tbody td {
border: 0;
}
.lightable-material.lightable-striped thead tr:last-child th {
border-bottom: 1px solid #ddd;
}
.lightable-material-dark {
min-width: 100%;
white-space: nowrap;
table-layout: fixed;
font-family: Roboto, sans-serif;
border: 1px solid #FFFFFF12;
border-collapse: collapse;
margin-bottom: 10px;
background-color: #363640;
}
.lightable-material-dark tfoot tr td {
border: 0;
}
.lightable-material-dark tfoot tr:first-child td {
border-top: 1px solid #FFFFFF12;
}
.lightable-material-dark th {
height: 56px;
padding-left: 16px;
padding-right: 16px;
color: #FFFFFF60;
}
.lightable-material-dark td {
height: 52px;
padding-left: 16px;
padding-right: 16px;
color: #FFFFFF;
border-top: 1px solid #FFFFFF12;
}
.lightable-material-dark.lightable-hover tbody tr:hover {
background-color: #FFFFFF12;
}
.lightable-material-dark.lightable-striped tbody tr:nth-child(even) {
background-color: #FFFFFF12;
}
.lightable-material-dark.lightable-striped tbody td {
border: 0;
}
.lightable-material-dark.lightable-striped thead tr:last-child th {
border-bottom: 1px solid #FFFFFF12;
}
.lightable-paper {
width: 100%;
margin-bottom: 10px;
color: #444;
}
.lightable-paper tfoot tr td {
border: 0;
}
.lightable-paper tfoot tr:first-child td {
border-top: 1px solid #00000020;
}
.lightable-paper thead tr:last-child th {
color: #666;
vertical-align: bottom;
border-bottom: 1px solid #00000020;
line-height: 1.15em;
padding: 10px 5px;
}
.lightable-paper td {
vertical-align: middle;
border-bottom: 1px solid #00000010;
line-height: 1.15em;
padding: 7px 5px;
}
.lightable-paper.lightable-hover tbody tr:hover {
background-color: #F9EEC1;
}
.lightable-paper.lightable-striped tbody tr:nth-child(even) {
background-color: #00000008;
}
.lightable-paper.lightable-striped tbody td {
border: 0;
}
</style>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to the pedtools package</h1>
<h4 class="author">Magnus Dehli Vigeland</h4>
<h4 class="date">2022-11-03</h4>


<div id="TOC">
<ul>
<li><a href="#pedigrees" id="toc-pedigrees"><span class="toc-section-number">1</span> Pedigrees</a>
<ul>
<li><a href="#the-ped-class" id="toc-the-ped-class"><span class="toc-section-number">1.1</span> The <code>ped</code>
class</a></li>
<li><a href="#basic-pedigree-plots" id="toc-basic-pedigree-plots"><span class="toc-section-number">1.2</span> Basic pedigree plots</a></li>
<li><a href="#builtin" id="toc-builtin"><span class="toc-section-number">1.3</span> Built-in pedigree
structures</a></li>
<li><a href="#examples-of-pedigree-construction" id="toc-examples-of-pedigree-construction"><span class="toc-section-number">1.4</span> Examples of pedigree
construction</a></li>
<li><a href="#pedigree-subsets" id="toc-pedigree-subsets"><span class="toc-section-number">1.5</span> Pedigree subsets</a></li>
</ul></li>
<li><a href="#markers" id="toc-markers"><span class="toc-section-number">2</span> Markers</a>
<ul>
<li><a href="#creating-marker-objects" id="toc-creating-marker-objects"><span class="toc-section-number">2.1</span> Creating marker objects</a></li>
<li><a href="#plotting-pedigrees-with-marker-data" id="toc-plotting-pedigrees-with-marker-data"><span class="toc-section-number">2.2</span> Plotting pedigrees with marker
data</a></li>
<li><a href="#markers-attached-to-ped-objects" id="toc-markers-attached-to-ped-objects"><span class="toc-section-number">2.3</span> Markers attached to
<code>ped</code> objects</a></li>
<li><a href="#accessing-and-modifying-individual-markers" id="toc-accessing-and-modifying-individual-markers"><span class="toc-section-number">2.4</span> Accessing and modifying individual
markers</a></li>
<li><a href="#gettingsettingmodifying-many-markers-simultaneously" id="toc-gettingsettingmodifying-many-markers-simultaneously"><span class="toc-section-number">2.5</span> Getting/setting/modifying many
markers simultaneously</a></li>
</ul></li>
</ul>
</div>

<p>The purpose of this vignette is to show how to work with pedigrees
and marker data in <code>pedtools</code>.</p>
<p>The following command installs the current CRAN version of the
package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;pedtools&quot;</span>)</span></code></pre></div>
<p>Alternatively, you may want the latest development version from
GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;devtools&quot;) # install devtools if needed</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;magnusdv/pedtools&quot;</span>)</span></code></pre></div>
<p>Now you should be able to load <code>pedtools</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pedtools)</span></code></pre></div>
<div id="pedigrees" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Pedigrees</h1>
<p>In <code>pedtools</code> and all related packages, pedigrees are
stored as <code>ped</code> objects. We start by explaining briefly what
these objects look like, and their basic constructor. If you are reading
this vignette simply to learn how to create a particular pedigree, you
may want to skip ahead to section 1.3 where we describe practical
shortcuts to common pedigree structures.</p>
<div id="the-ped-class" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> The <code>ped</code>
class</h2>
<p><strong>The <code>ped</code> constructor function</strong><br />
The most direct way to create a pedigree in <code>pedtools</code> is
with the <code>ped()</code> constructor. This takes as input 4 vectors
of equal length:</p>
<ul>
<li><code>id</code> : individual ID labels (numeric or character)</li>
<li><code>fid</code> : id of the fathers (0 if not included)</li>
<li><code>mid</code> : id of the mothers (0 if not included)</li>
<li><code>sex</code> : gender codes, with entries 0 (unknown), 1 (male)
or 2 (female)</li>
</ul>
<p>In other words, the j’th pedigree member has label
<code>id[j]</code>, father <code>fid[j]</code>, mother
<code>mid[j]</code>, and gender given by <code>sex[j]</code>.</p>
<p>For example, the following creates a family <em>trio</em>,
i.e. father, mother and child:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ped</span>(<span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">fid =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">mid =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>), <span class="at">sex =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  id fid mid sex</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   1   *   *   1</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   2   *   *   2</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   3   1   2   2</span></span></code></pre></div>
<p>In this example the child (<code>id=3</code>) is female, since the
associated entry in <code>sex</code> is 2. Note that missing parents are
printed as <code>*</code>. Individuals without parents are called
<em>founders</em> of the pedigree, while the <em>nonfounders</em> have
both parents specified. It is not allowed to have exactly one
parent.</p>
<p>Instead of numerical labels as above, we could have used character
strings. Let us create the trio again, with more informative labels, and
store it in a variable named <code>trio</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>trio <span class="ot">=</span> <span class="fu">ped</span>(<span class="at">id =</span> <span class="fu">c</span>(<span class="st">&quot;fa&quot;</span>, <span class="st">&quot;mo&quot;</span>, <span class="st">&quot;girl&quot;</span>), <span class="at">fid =</span> <span class="fu">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;fa&quot;</span>), <span class="at">mid =</span> <span class="fu">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;mo&quot;</span>), <span class="at">sex =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>trio</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id fid mid sex</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   *   *   1</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   *   *   2</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl  fa  mo   2</span></span></code></pre></div>
<p>The special strings <code>&quot;0&quot;</code>, <code>&quot;&quot;</code> and
<code>NA</code> are all interpreted as a missing parent.</p>
<p><strong>The internal structure of <code>ped</code>
objects</strong><br />
From the way it is printed, the object <code>trio</code> appears to be a
data frame, but this is not exactly true. Rather it is an object of
class <code>ped</code>, which is basically a list. We can see the actual
content of <code>trio</code> by unclassing it:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unclass</span>(trio)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $ID</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;fa&quot;   &quot;mo&quot;   &quot;girl&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $FIDX</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0 0 1</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $MIDX</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0 0 2</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $SEX</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1 2 2</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $FAMID</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;&quot;</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $UNBROKEN_LOOPS</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $LOOP_BREAKERS</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; NULL</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $FOUNDER_INBREEDING</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; NULL</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $MARKERS</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>In most cases it is not recommended for regular users to interact
directly with the internal slots of a <code>ped</code>, since this can
have unfortunate consequences unless you know exactly what you are
doing. Instead, one should use accessor functions like
<code>labels()</code>, <code>getMarkers()</code> and
<code>founderInbreeding()</code>. The most important accessors are
described within this vignette, while others are documented in the help
page <code>?ped_utils</code>.</p>
</div>
<div id="basic-pedigree-plots" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Basic pedigree
plots</h2>
<p>To plot a pedigree, simply use <code>plot()</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(trio)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAATlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6kNtmAABmADpmtrZmtv+QOgCQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///8NKk6TAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAC/UlEQVR4nO2cDXOiMBiEI63SespdS9Hz///RI3KnbSUJJHmzeW92Z+qoyG4e8gkzjbkol0EXIFUEQIsAaBEALQKgRQC0CIAWAdAiAFr/JYDxqWDRFkXOAqz8vYQ+l9uLUCmA8X4MHcIDPKS4K6FKgDWFqhFgPsORXCHAipI6vsYCrGrtBBCQO2H2SHUAOSYtAqSIACEPcUkDeI7l+H11TYgAoWPVAdQ2kREgdIAAscVZfQrofiBm5l7c1l3fowFmb+FVATyehnsqEbv4+1pij0utAPczYxZ4kYvFpf5LT10UKL2+h94/aPAnANqfAGh/AqD9CYD29wb0xuwk/bPIE3Buk4uPBtiL+ueRO+C0NaZ5v3TjgjABJAbgtH1tx9ZrC2CTB+NryqEa6J7HvrA5RhQj6O/UaTsG9vbq9ePfMFKc2+eIAAtwfjlaw/gqiAPY3V72v3/Yqz84r+GCPjCktKE4gP39Zbp67msYAhiH0uZX8Rp4AHAPKAGAu1ek0DUwjL3o2o1ilQyQ2AeszblNmJCTARJHIdsHNscufhxNB4ifB7KIq1F0AAHQAQRABxAAHUAAdAAB0AEEQAcQAB1AAHQAAdABBEAHEAAdQAB0AAHQAQRABxAAHUAAdAAB0AEEQAcQAB1AAHQAAdABBEAHEAAaEPefWyszBK3N7Y1kipixcX/KmlPKt1iQmK1QJRR0VQTgMC2ZJeNZNEzCUwmAx7JsmoAlAUpYeh0Lx+V3JADakQAFHNWPQgTIlpbfU8laSD+A+vsBl6tyAFX3xDO2yp5KfPfV91zo685SGp/MXc0j97ValSHoXUQEQIsAaBEALQKgRQCvbluiZNlsaVaFaoAATgkBdMZsDpvjtMfSwTRvygC669ZSfwHapw9tNTDtBtT9A9ipa0LXXYnsnkATwF4dQK8dQH0NTH2g1wvwbRRSCGDngeZn864XwGrqCdKS6wNZdvsMS6YG7MZkKXtkrhDvB9AiAFoEQIsAaBEALQKgRQC0CIAWAdAiAFp/ABimzvgMA6zZAAAAAElFTkSuQmCC" style="display: block; margin: auto;" /></p>
<p>Under the hood, <code>pedtools::plot()</code> is an elaborate wrapper
of the excellent plotting functionality of the <code>kinship2</code>
package. Most of the possibilities provided by kinship2 are available
from pedtools, and several features are added. An overview can be found
in the documentation <code>?plot.ped</code>, but a quick example should
get you started:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(trio, <span class="at">deceased =</span> <span class="st">&quot;fa&quot;</span>, <span class="at">starred =</span> <span class="st">&quot;mo&quot;</span>, <span class="at">hatched =</span> <span class="st">&quot;girl&quot;</span>, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="at">title =</span> <span class="st">&quot;Trio 1&quot;</span>, <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAkFBMVEUAAAAAADoAAGYAAP8AOjoAOmYAOpAAZrYA/wA6AAA6AGY6OgA6kJA6kNtmAABmADpmOgBmkLZmtrZmtttmtv+QOgCQOjqQZgCQkGaQttuQ2/+2ZgC2Zjq2kDq2tma2tpC225C2/7a2///bkDrbtmbbtpDb25Db2//b/9vb////AAD/tmb/25D//7b//9v////dYIyLAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAEdklEQVR4nO2diZLaMAyGDWl3Q1voBS3d9E6v3XC8/9s1jiFctogdKUIz+mfqgTUb6bN8xNqYmq1wGW4H+koBuKUA3FIAbikAtxSAWwrALUSAwuw0bd6uJuOfvg9N8SxaDQzwy9wwwNY6PXoAqjefjQiA9Wy0nIy/NRFYvTHm+SdXu1mYZy9lANQNfffHAlSuU71vajcf3v5bCAG4f/xrx0DzavPFtKNhIwVg7gaxG8jufSMpALa0zldN2x95LQ1AfATEjoE9wH4W2gdAHsD299E6sL19AAYpALcUgFsKwC0F4FYCQI7vRQ/FA9yW//EAN+a/FyCHNJxrT3uBn/IChD8+IMDBbxAhDiAfqgedtTtAEAWQDzUELhwOByEGIAfqUOVzN0QQAZADdajy+tofIAfqUBVwNfDjzgA5UIeqYG/3V3QFyIE6XNEA5EAdssJTpremG0AO1GELmPN9VZ0AcqAOW9CqmwqQA3XoIgDIgTp84QOcvjPkwgaAwtFB0fsN7AhA46GLuAGgCamTmAGgFaGbeAGgJbmj4pMGiCsxdE/UVYgA0fdC0E1pZzECQLuC7kpInCHtB6BtWYRSUpd+VyMBoH1xjJJyrz5fI7eUUGIiSmnJ495ZCSgzFKfE7PcT+PaaASj9NhDASWYrPjOHmBvt8feH5Nwo6i0/+R9QLg3gblmGB0DecjFEQNb1FYD7+grAfX0F4L4+aKBEeFIbDaC8jzawniE8n4cEUI2Wr2ZeBBBgHq7sKqwIbBa+4xSggdXE2AdWi+PHPuMVDbCavJvVXddat2ar3YmD6u6jvw9di0BR/1oJnmqAlQBQWytt05XNg7Pz2o/ah2K+euENwRWA9evmEdb0ECQATNtivlk0rQ81YIcxUPXpQwkA80Phmg5swGsA9VQ6/j5sBC4AwNnkCsDhconijkBlJ6+KrQv1HwP2SutZjwW5H8BhFkoxsB8Do4cifR7tCdCuA3gGIqV3o9wGFIDbgAJwG1AAbgMKwG1AAbgNKAC3AQXgNqAA3AYUgNuAAnAbUABuAwrAbUABuA3IBsiyzNT/KE1QAljPM/eCkIEKwDntHM8cDI2IAJzDrf9tiS8aAJ//REEgAmiLo1IQQMB/EgIKgGD7UxBQAXj7jxCAoP8kBDQAIf+lAATbXwTAif/ZWf/BJyAAOPa/LciGMUUE2vLSfzEAofaXAuC/lctkdKF9Ow/jPwnAuf+HHwgBaIvAS1QR3QtdTEWS7oXOlrK2oNnS0O0HhvGfcEd24b+cDc1hFJy0v6At5cmgpew/W8K80HYY/+kyc9lZaotK1LlRJ0Ib5NljaikAtxSAWwrALQXglgKAao+woZyM9WqgCChAUEQAhTGj5ejBHYhdmvEPYQBFcxR4BzC7e5QWAXd6s9gDTMV1oeYIqT3D6QDm4gBK6QDiI+DGQCkX4GwWEghg14HxV/sfqkkFsKoC38WBK7oxgPLVLNdFE4HK9PtCkwjpfoBbCsAtBeCWAnBLAbilANxSAG4pALcUgFsKwK3/yPPTA3tUFfUAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>See Section 2.2 for how to add, and control the appearance of, marker
genotypes to pedigree plots.</p>
</div>
<div id="builtin" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Built-in pedigree
structures</h2>
<p>Rather than using the <code>ped()</code> function directly, it is
usually quicker and safer to build pedigrees step by step, applying the
arsenal of utility functions offered by <code>pedtools</code>. A typical
workflow is as follows:</p>
<ol style="list-style-type: decimal">
<li>Choose one of the basic pedigree structures as starting point.</li>
<li>Add/remove individuals as needed.</li>
<li>Modify attributes like genders and labels.</li>
</ol>
<p>You will find several examples below, but first let us list the
available tools for each of the 3 steps.</p>
<p><strong>Basic pedigrees</strong><br />
The following pedigree structures serve as starting points for pedigree
constructions. For parameters and details, see
<code>?ped_basic</code>.</p>
<ul>
<li><code>singleton()</code>, a pedigree consisting of a single
individual</li>
<li><code>nuclearPed()</code>, a nuclear pedigree
(parents+children)</li>
<li><code>halfSibPed()</code>, two sibships with one parent in
common</li>
<li><code>linearPed()</code>, a straight line of successors</li>
<li><code>avuncularPed()</code>, uncle-nephew and similar
relationships</li>
<li><code>cousinPed()</code>, cousins of specified degree/removal</li>
<li><code>halfCousinPed()</code>, half cousins of specified
degree/removal</li>
<li><code>ancestralPed()</code>, a family tree containing the ancestors
of a single person</li>
</ul>
<p>There are also more specialized structures, including double cousins,
selfing pedigrees, and consecutive matings between full siblings. Look
them up in <code>?ped_complex</code> if you are interested.</p>
<p><strong>Add/remove/extract individuals</strong><br />
The functions below are used to modify an existing <code>ped</code>
object by adding/removing individuals, or extracting a sub-pedigree. For
details, see <code>?ped_modify</code>.</p>
<ul>
<li><code>addChildren()</code>, with special cases <code>addSon()</code>
and <code>addDaughter()</code></li>
<li><code>addParents()</code></li>
<li><code>removeIndividuals()</code></li>
<li><code>branch()</code></li>
<li><code>subset()</code></li>
</ul>
<p><strong>Edit labels and attributes</strong><br />
The following functions modify various attributes of a <code>ped</code>
object. See <code>?ped_modify</code> for parameters and details.</p>
<ul>
<li><code>setSex()</code></li>
<li><code>swapSex()</code></li>
<li><code>relabel()</code></li>
</ul>
</div>
<div id="examples-of-pedigree-construction" class="section level2" number="1.4">
<h2><span class="header-section-number">1.4</span> Examples of pedigree
construction</h2>
<div id="example-1-trio" class="section level3 unnumbered">
<h3 class="unnumbered">Example 1: Trio</h3>
<p>As our first example we will recreate the <code>trio</code> pedigree
without using the <code>ped()</code> constructor. To give a hint of the
flexibility, we show 3 alternative ways to code this.</p>
<p><strong>Alternative A</strong><br />
The obvious starting point is <code>nuclearPed()</code>, with
<code>nch = 1</code> to indicate 1 child. By default, this creates a
trio with numeric labels (father=1; mother=2; child=3) and a male child.
Hence we fix the gender with <code>swapSex()</code>, and edit the labels
with <code>relabel()</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>trio2 <span class="ot">=</span> <span class="fu">nuclearPed</span>(<span class="at">nch =</span> <span class="dv">1</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>trio2 <span class="ot">=</span> <span class="fu">swapSex</span>(trio2, <span class="at">ids =</span> <span class="dv">3</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>trio2 <span class="ot">=</span> <span class="fu">relabel</span>(trio2, <span class="at">new =</span> <span class="fu">c</span>(<span class="st">&quot;fa&quot;</span>, <span class="st">&quot;mo&quot;</span>, <span class="st">&quot;girl&quot;</span>))</span></code></pre></div>
<p>With the pipe operator <code>|&gt;</code> the above commands could be
written in one chain, like this:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>trio2 <span class="ot">=</span> <span class="fu">nuclearPed</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">swapSex</span>(<span class="dv">3</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relabel</span>(<span class="at">new =</span> <span class="fu">c</span>(<span class="st">&quot;fa&quot;</span>, <span class="st">&quot;mo&quot;</span>, <span class="st">&quot;girl&quot;</span>))</span></code></pre></div>
<p><strong>Alternative B (quickest and best)</strong><br />
Even quicker than the pipe version is the following one-liner, where all
details are specified directly in the call to
<code>nuclearPed()</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>trio3 <span class="ot">=</span> <span class="fu">nuclearPed</span>(<span class="at">father =</span> <span class="st">&quot;fa&quot;</span>, <span class="at">mother =</span> <span class="st">&quot;mo&quot;</span>, <span class="at">children =</span> <span class="st">&quot;girl&quot;</span>, <span class="at">sex =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><strong>Alternative C</strong><br />
Here is another possibility. We start by creating the father as a
singleton, and then add the daughter:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>trio4 <span class="ot">=</span> <span class="fu">singleton</span>(<span class="st">&quot;fa&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>trio4 <span class="ot">=</span> <span class="fu">addDaughter</span>(trio4, <span class="at">parent =</span> <span class="st">&quot;fa&quot;</span>, <span class="at">id =</span> <span class="st">&quot;girl&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Mother: Creating new individual with ID = NN_1</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>trio4 <span class="ot">=</span> <span class="fu">relabel</span>(trio4, <span class="at">old =</span> <span class="st">&quot;NN_1&quot;</span>, <span class="at">new =</span> <span class="st">&quot;mo&quot;</span>)</span></code></pre></div>
<p>Note that <code>addDaughter()</code> automatically created the mother
as “NN_1”, so we needed to relabel her.</p>
</div>
<div id="example-2-an-inbred-child" class="section level3 unnumbered">
<h3 class="unnumbered">Example 2: An inbred child</h3>
<p>This time we will create this inbred family:</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAS1BMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6kNtmAABmtv+QOgCQZgCQ2/+2ZgC2/7a2///bkDrb////tmb/25D//7b//9v////qHGfVAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAF/UlEQVR4nO2dbXfaRhQGlaRua+qWlgrM//+lBZPk+NRGs293WcHMp8Tavc/VsAgEiXc6yiLTrRsYHQUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAcuCph+0zVwkJqq8wmLxd38uTsjKbLqkm5zA1Xn/r9iy8eKDWSltTuDatI8/b7eIughqdQJXJn3647je24c0O4EcQaHNd8koKJ5VaD2C2p1A3lIM7b5hRMMTUFBBqaX3RtkBuUVaJLQ8gUxB0f3HB+TWV1BBLQXBeAXB+MX6LVhsKDjg9iuockR0gIJghIJgRLCghPJjC4p+o6ig5Sop5QcQNPbN6iMIqvo4ZQRB0Z8H5S6szFEdBNWcQNqc3Jf/nFE9BFWcQOKcD+/XU78UGERQ+Qkkp00LfyurmDKimaDiE0hP+6mcbjAzG+klqPAEctIS7r4zK76rWnCrnxjwISpjRk3arSr2DghOU1D3ir0DgtMU1LngTSLiwro03/ifoS1nDV3u5jmrFbTWoPUu/T45q756lsbsf/unWa0PvL6c7gOes6d9j0vJ203Tl78KAy4pcPyw+ZoqqMTPqfnd9Ev2xOTE3an7ucoQZMzTlCqoYMnvn/44Xk6iEH54n88PQ/EjgBHz9Dwntl9+Sah5hFNSIwWdSBRUccnclq+gpNxd5FOsg6C5+Cp9CeaLRF19GpAmqMZPzRMgJfv15du/cdXTBN1s/SSlV72MNRFU7mfXwA/mX14sY2onCarxU9H6+w6utHBRk/pC/HlpGsDVy/3sn1qsn6UmtqfLz+tLTUq9oIoX+N3lO4e6e4HlNran+lWrtPqWr+8HoAsENXI/goI6qa06jp+gT2Ira47k5xjSTl3JwfyM9y2WgkJnB6AgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIeDhBTX5RRbf/OtN69gABCqqur6DQDhQUOnuAAAVV11dQaAcKCp09QICCqusrKLQDBYXOHiBAQdX1FRTagYJCZw8QoKDq+goK7UBBobMHCFBQdX0FNe6g7kum/KzGJfNPIFtQ4bFsvrfbXFLuCYwq6H2tpo7uQ9DHXzParnTmsSEFfVIosvbisREFfVontPjCsQEFXSkTW/3qMQXBsfEEXa0SXP7KseEELRSJrq8gKqIgKKIgKLIGQYs1ogPCBUX/s9PVC6qanTBCQfcuKPg59gAX6YQ0BZUPURAMUdDymEe4WU1Ke+SPOxRUMRtGNfJzF4I+G9bKz30IetyvfdLT3o+EW9g87kXQer96Lv2ooiCtuZwfRXNOoH0Hi911TWuCggAFAQoCFAQoCFAQoCBAQYCCgLqWt5m75mSlHTbnd/+1O/8ssH9KKF8laJ4iBe1/bbHfxnXmU/OHDRmqEXR6iCMFVe04w1w2bMHt8WoE7b79GSloF/jsOiYv0ApBp4TQa9D296ly87RF5q9/bxLqlws6L9FIQYfNufg2zNDuvL0l739YLmh36j90Bb0RdyG6bH0Yt7fT21M4XlDV5mmLXC7PWL9Y0Petr/L2pioQFPZaf1k7WH/cN4oNdpdb5LBJqj+uoOP2fP2Mu0i/XUR597qBBdXvLgfMKW8jvFkFFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEFAdMulv1x7GMIF3TK8BQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAT43Tywhh5vioIABQEKAhQEKAhQEKAgQEGAgoBuguaSjUoGoO8K4v3OhqOvINzubDy6Cto/xe2zEkVXQdv1LaCugg6b9S2groLmL7HbXobQU1DuVklD0FHQKp9hPQXF7cYXSUdBq7wE9RS0wneJR29WEQUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCDgP9/f0TMom1PuAAAAAElFTkSuQmCC" style="display: block; margin: auto;" /></p>
<p><strong>Alternative A</strong><br />
One approach is to first create the pedigree consisting of individuals
1-6, with <code>halfSibPed()</code>, and then use <code>addSon()</code>
to add the inbred child.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">=</span> <span class="fu">halfSibPed</span>(<span class="at">nch1 =</span> <span class="dv">1</span>, <span class="at">nch2 =</span> <span class="dv">2</span>, <span class="at">sex1 =</span> <span class="dv">1</span>, <span class="at">sex2 =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addSon</span>(<span class="at">parents =</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">5</span>)</span></code></pre></div>
<p><strong>Alternative B</strong><br />
We could also view the half siblings 4 and 5 as half cousins of degree
0. The <code>halfCousinPed()</code> function accepts an option
<code>child = TRUE</code> adding an inbred child. The labels will be
different with this approach, so it might be useful to plot intermediate
results to see who is who.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">=</span> <span class="fu">halfCousinPed</span>(<span class="dv">0</span>, <span class="at">child =</span> T) <span class="sc">|&gt;</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addSon</span>(<span class="at">parents =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">|&gt;</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relabel</span>(<span class="at">old =</span> <span class="fu">c</span>(<span class="dv">7</span>,<span class="dv">6</span>), <span class="at">new =</span> <span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">7</span>))</span></code></pre></div>
<p><strong>A note about the order of pedigree members</strong></p>
<p>Although both <code>x1</code> and <code>x2</code> reproduce exactly
the plot shown above, they are not identical objects:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(x1, x2)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>The reason is the <em>order</em> in which the individuals are stored.
For <code>x1</code> the ordering is the natural sequence
<code>1,2,3,4,5,6,7</code>, but for <code>x2</code> our construction
process has produced a slightly different order:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>x2</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  id fid mid sex</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   1   *   *   2</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   2   *   *   1</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   3   *   *   2</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   4   2   1   1</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   5   2   3   2</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   7   4   5   1</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   6   2   3   1</span></span></code></pre></div>
<p>The internal ordering is usually of little importance in
applications.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> However, if you get annoyed by “wrong”
orderings such as for <code>x2</code> above, you can use
<code>reorderPed()</code> to permute the pedigree any way you like. In
fact, the default action of this function is to permute into the natural
order of the labels, which is exactly what we need to make
<code>x2</code> identical to <code>x1</code>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">=</span> <span class="fu">reorderPed</span>(x2)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(x1, x2)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div id="example-3-a-complex-family-tree" class="section level3" number="1.4.1">
<h3><span class="header-section-number">1.4.1</span> Example 3: A
complex family tree</h3>
<p>Here we consider the family tree below, extending both upwards and
downwards from a single person. We will use this example to demonstrate
the <code>mergePed()</code> function, which “glues” together two
pedigrees by the indicated members.</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAVAAAAFjCAMAAABc2KPoAAAATlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6kNtmAABmADpmtv+QOgCQZgCQ2/+2ZgC2/7a2///bkDrb////tmb/25D//7b//9v////XrGT5AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALBklEQVR4nO2d62LaSgyE3fb0Es6hoXUd4P1f9PhCGpKALe2OVtrNzK+2NqPRh2/YoHZnCqrOO0BrIlCwCBQsAgWLQMEiULAIFCwCBYtAwSJQsAgULAIFi0DBIlCwCBQsAgWLQMEiULBuAe3WhKqL9VuNjK6xvpbw3yTLxLoOhel21QVR4lVkLSBjoG/yQDYga6Dd6l+3FtkCfe+Abxjt/87h/lZQHOjtg7aFq2xhqvs90xBA821NgaoilwaqiAGwhZjrIhcGqjr0IHzzvZWRCTTZOgJQ5eUGxNjO+uYSAs1wlsIjUOGL3YH6tE2g1ThDgGYpNfmm1p2tMhfYQvW3tqTOGXUzQrnv8gRaFGjIszyBql+dBdT5wp5AywKN+FkeHJlA6wZ6b4VMnpZAlZFLA5UfeFSyBKqLjAaaFM/zmZKk9q11CgE9n9c/rC1raD23lQh0O+wdi7JPPXXfrvB6Lr+Z8s3awnIWQJcAqy6daDW5tEAT6l5FVn96Sr1h9M5mdaHObbOWPHNiXVlkWEv3Mpj6J8g6knnDoZgWCFOiW+S+HT5FqU6dmZYrX7BNN6ZFC5ft0YFp6ZJtN+jxBpYueC51enA6FXqdKoy79TsHOp577Xr2vKLYLP30/bdF3dN+3IgeLJz70fjRwnjS4cufjTW2gB53ny2AnvafHscN6R+8cz/m7T4ZER26XKBD15kAffr631i7x3sfd+Nm3+0N3qnZPBfo0D0MJkCX2oPRlmQFtP/yM3uXP1sCPRh5G+3yT98e84+hpkBtzkqDke9p/wA4KRkCHSzOSbO6/WbfCepH08hAB8PrUIuD87jDIy6bzID23YMd0PkiAqz+8ghkw9kLaD8GswC6XI+Z7VZht9Cnrw9GH3unlru9yVnpHBjoZf+xuLw5bO+WGeb5QA1ldwy1Mv6otZtsyrN2k0151m6yKc/aTTblWbvJpjxrN9mUZ+0mm/Ks3WRTnrWbbMqzdpNNedZusinP2k025Vm7yaZuFgP9HEJSBeWmrh6mGOj3Nc8ovZi2BfTNz7Oa/zqjNdB3Fg5EWwJ66xdz2aaAEE7FspPcNGj7K+GmQO+8vjTR5oGWJtoM0LsvJ1C0dcO/UyLQksWykqy9mEDBzgQKdv6wQLOUXBatOMXWk2zk5BaqXEig+mIECi5GoOBiWUB5Ya9cSKD6YnlAeXNEt1CQk/dDNQsJVF8sFygfgWgWinLyIZ18oSwnHyOLFwpzXt8k+Qhzm8yBnp+/gfNB5jZlfFlMkdPzq2IuR5kk1ZKzmqC15KwmaC05qwlaS85qgtaSs5qgteSsJmgtOasJWkvOaoLWkrOaoLXkdAt63E2fEOXD7+Q5n76aDHqVygvoPElOIXHO4cuf8d3yI+oFVDtfS5rzNA8VMxidK5UX0F65DUlzard8uLyAHn50qrmp0pzD5187o4GsMjkBPe6mIXIHeePSnP00hPtkNINZItfLEcWBVAx0nk9oOOd4S65AFXNTxUBnlBYDWYXyBSo/g8iPob91xmg5AV02IYNd/rhTGqPldpafThsGJ6V5lvfJbMLtttx2eeUgWnnOQXc9hlYtNx1qyVlN0FpyVhO0lpzVBK0lZzVBa8lZTdBaclYTtJac1QStJWc1QWvJWU3QWnJWE7SWnNUErSVnNUFryVlN0FpyIoOCR9Vm/MJBXwNjN1uifK5CYTwzfoMjLHAdGYYU9Va//huk3+SFSf6wjd7GBRDPGOh7hwI7Vo5JtrEt0FvvuBkLiEdsoLdfb36kyrKw6Rljfu/ldjQQDkZNA7zvv9r40J/nQKBoA6uuDa0JFG1tey7NNIgKdO3FoYHatU2gYOMPCzRLq1WtnIMDzamdsYVm1CVQcN3QQLe8CVRpEBVos9ehAYEG+KSU8VmeQHUOBIq12HZ2Axr7fugdD4GxH9A7a5jBQJjEBhr7mdINF9FTT0+gN9YxQoGwET5EdgX6dqVYz+XPSd8c8QUa+5sjs5XkNpG8dvqtKHlT6sgCS5xVnNpNNuVZu8mmPGs32ZRn7Sab8qzdZFOetZtsyrN2k0151m6yKc/aTTblWbuNpobLpzj5qBtx7V5lq2lqdP6EnZkFfjNVo1Cltad5doPJTKLD6Awe6wYGqpqEKqy9zFo8GLxTx908vXUaFAkTFujTV80IKnegwzLKEbrTY4EeVDPn3Hf5C1DV8RlVW6RlF8LXHpSnDqnxcvjs4wJV7jzS2tN2rzqYKE9KcYEedId3Ye3LdqQ4msibOnTd51//Rj2GKvd4aW39qUPX1NN35LBRJFDtFZ1qC9XMWNU1FfeySXv94X4MnT+GYK+aoEC1/7+B/NzR6Wasio2n/5Ek9kfPGLWbbMqzdpNNedZusinP2k025Vm7yaY8azfZlGftJpvyrN1kU561m2zKs3aTTXnWbrIpz9pNNuVZu8mmPGtHayr91wKawtifCoTJfBOocn29LsFwza7aoDOrk5gDvTbBMDUHep1yNbID0HdxEJ7JC9P87yMtD/SGg4kp0P4GvjhAbxrYuKLcVZFLA73zeiNbiLkuMoGmescAqjr0IHzzvZWRywJVXm5AjO2sxcdWApW9mkDRzlJ4BCp8sTtQn7YJtBpnCNAspSbf1LqzVeYCW6j+1pbUOaNuRij3XT59sdtJiUDVryZQveyAOl/YE2hZoBFvjoAjFwZ6d4W4QJWRgwDN5OkBVPHPlkDlBx6VLIHqIhcHemuVbJ62QD0f0knQBHuMLKr+LnIkoG/W8v6igzDy68w6O3OgL9+9QH0XxxzoS9SUGyapN4w0e6/ETa70m1yaA44kMqwlE7sSik2AQGPblVBsAgQa266EYhMg0Nh2JRSbAIHGtiuh2AQIdNXuMnNLMctPmu7vNC/sWK9nv9O+Ew8mUkbemom7ZnfczYOYpklR0mlRwnQX56s/gLT4nfZj3L6TTRzVRZ5mU63CWLEbN8wl3fRWC+ehytJdnK/+ANLFTzV+UBV5Hu+3TIhV2w3dw2CT7tn55Q8gvfYT7lW6yJMSgZ4vA/yevj2exbP8pAekv3ZQoK/9hON21ZHnibtpdrOHatpkIKCD8KykjTysn+6aBToIz0kJW+hpvzIgs9VdXrp9pgBd3boEQPGnTHugvXw8ZgLQtcmz20ANLpvMgWoGf6sib0/b3QZqcWFvDNRkHO6SdJrbvWxiKXZ/9x/4R09boP3ysNPkPDpN213b/HlzJDYBAo1tV0KxCRBobLsSik2AQGPblVBsAgQa266EYhMg0Nh2JRSbAIHGtiuh2AQINNsu4xcOXjKNbGoQFWjismzzbAMCBRsQKNiAQMEGBAo2IFCwAYGCDQgUbECgYAMCBRsQKNiAQMEGBAo2IFCwAYGCDQgUbECgYAMCBRsQKNiAQMEGBAo2IFCwAYGCDQgUbECgYAMCBRsQaIoBv9v02hyRkHoRgYJFoGARKFgEChaBgkWgYBEoWAQKFoGChQP6Mlr1sDLJLJSWyMfd9JFTONhtUzCgL6NVh64SoJfIyyw6lFBAX0arjm94HUCfI2NHR4GAXg0s7b/8rALo38g9am+fhTuGXoCOO1A1x9DL9LUf8gnY20IDnebC1QX0uJviHlBE0UD7MV1dQN/9MU9goPMZs0agayNBVQIDvcxFVIzv9NQ1UNS1E/ykdK5tC92esaoSgS7De8OelCZVBnRzxqpKvDkCFoGCRaBgEShYBAoWgYJFoGARKFgEChaBgkWgYBEoWAQKFoGCRaBgEShYBAoWgYJFoGARKFgEChaBgkWgYP0PRG1p0/53IaAAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Top part</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">ancestralPed</span>(<span class="at">g =</span> <span class="dv">2</span>) <span class="co"># bottom person is &quot;7&quot;</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Bottom part</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">halfCousinPed</span>(<span class="at">degree =</span> <span class="dv">1</span>) <span class="co"># top person is &quot;2&quot;</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">swapSex</span>(y, <span class="dv">4</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Changing sex of spouses as well: 5</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>z <span class="ot">=</span> <span class="fu">mergePed</span>(x, y, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;7&quot;</span> <span class="ot">=</span> <span class="st">&quot;2&quot;</span>), <span class="at">relabel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Note the argument <code>by = c(&quot;7&quot; = &quot;2&quot;)</code>, which means that
individual “7” in <code>x</code> should be identified with “2” in
<code>y</code>. As seen in the plot below, this individual ends up as
“8” after relabelling in the final result.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPedList</span>(<span class="fu">list</span>(x, y, z))</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA2AAAAFQCAMAAADuqe0SAAAAYFBMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6OpA6kNtmAABmADpmAGZmZjpmkJBmtv+QOgCQOjqQOmaQ2/+2ZgC2/7a2///bkDrb////tmb/25D/29v//7b//9v////YL3AhAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAUPElEQVR4nO2d60IT2RJGW0cYZUaPotFABnj/tzwEvEBS/VXt+w6s9cORJLuqurpWuulOnOUOAJqxjC4A4CWDYAANQTCAhiAYQEMQDKAhCAbQEAQDaAiCATQEwQAa8kSwBX5jtgp+Q380tmB1lH0J2APUu4p5oT8aBHNggDT0R4NgDgyQhv5oEMyBAdLQHw2COTBAGvqjQTAHBkhDfzQI5sAAaeiPBsEcGCAN/dEgmAMDpKE/GgRzYIA09EeDYA4MkIb+aBDMgQHS0B8NgjkwQBr6o0EwBwZIQ380CObAAGmS+yO/0RFPm72ycwEI5oBgmnTBkl69EmM5/EvK4q4FIJgDgmlGCLas/D19dfsCEMwBwTQDBFvET8nLWxeAYA4IphkuWOnyxgUgmAOCafoLdvjy7oIlFYBgDgimGS9Y8fqmBSCYA4JpugtW34+mBSCYA4JpEAzBikAwTQ3Bcv+hwUC24QUgmAOCadodwcIPtzqCVSkAwRwQTINgCFYEgmm6C9bvKmKVAhDMAcE04wVrdh+sSgEI5oBgmv6CHT7RXbCkAhDMAcE0wwVL3hd9C0AwBwTTDBDs2TPpu6JvAQjmgGCaEYJ1+j5YlQIQzAHBNOmCRb9QrH+1yf4qcucCEMwBwTTt+jO8ybW3AsEsEEzTULDBXa6TH8EcEEzTsj8jFauVG8EcEEzTtj+jFKuXF8EcEEzTuj8jFKuZE8EcEEzTvj+9FaubD8EcEEzToz89FaudC8EcEEyT3Z+bjz8SshxGvP2yLGfx9QabZfns51FsLwIvQjAHBNPk9ufmn7cJgh2N/vbs3rFjQeJcnx0pnnjjeneOYBVAME1mf26/fk85gj0EPZz/0BFE8Fyw1HPD28tPCFYBBNP0OUX8Ffdp4JwAT9k+PUVM/9Vre8EpYg0QTNNVsKci3Pz7LWP9Mza/DMu4srH7cIVgNUAwTWfBfsuw+7vYr1/nmFkXDrf7j/oGDEMwBwTTdBfsUYjd+7Lzw/1FjscjWP51eY5gNUAwzQDB9lJsggeQdTYP60tueyFYDRBMM6g/lT7q3qFQ76+vHQTTDOtPuRxdPiCCYA4IphnXn3LBalSRkATBLBBMg2DhJAhmgWAaBAsnQTALBNMgWDgJglkgmAbBwkkQzALBNAgWToJgFgimQbBwEgSzQDANgoWTIJgFgmkQLJwEwSwQTINg4SQIZoFgGgQLJ0EwCwTTIFg4CYJZIJgGwcJJEMwCwTQIFk6CYBYIpkGwcJKgYOH/f5nOW760TwGNBii/cmPzO++A53HCD9bmxQoWeMTN+ntJzoR1LaDJAJVsv7Gi8w7wcyOYlaSfYE8XpL+H9i2gxQAVbX8Nwcp2gJ8bwawk3QRb5I+pyxsX0GCAFvFT6uoKIRCsJQjWXbDDxYnBygUrLCCwGMGsJL0EO3p58YS1LOAVCFa0OQgWToJgsWzpCZ21adGKBatxkchZi2BWkk6C1fejaQEIlhqtMGBZ5q4BEpMgWOzVyQlL0odyI9iYAIlJ8gVLIxAxWnKHAnoIVlh+7x3gr0UwK0ntI1j44VZHsCoFjDuCxR/uvAP8tQhmJUGwcBAE02tbjm7omB6NU7k2K433V7VCv3otyNHjnQVLK6C9YOEyVx9+XYJVyfbTrfaOTSBY6ib2LaD+AEXTtxKseAf4i2cX7IlWL1Cwwye6C5ZUQHPBolWKh8sEKxuxUxRsWf2hPkMEK9vAvgU0GKBg+maCle4AP8ncglV9g0lI1k2wcKoZCmgyQKH07QQr3AF+kpMSrK1hYwQr+h2zbwFtBugxrb6O1VKwer/kn55gRy+fTbDoZVLn1Kvs+mq3AloNkJ99bauG7wA/yWkJ1qvceqcNFYMML6D7ANVLMq5KBLNiI1g4CILpJJ0FS2RUuQgWDoJgOskUR7C1IhCsFgjWAQTTIFhOEATTSRDMil1bsNJ/r2iO7AiWkWRqwWa/D5YUfZhi1TIjWEYSBLOCJwp28/FHJP5KqNsvy3IWWL/OZlk+p2Z9zvYi8KJmA+T2TyaJ9E9XqfqXwAkKdvBM2/ehbMFu/nkbEWwt1vbsfkZKdvD12eqIBo9eu/ORgvn9k0ki/ZMBRP+SOEXBno164+N8rmC3X79Hd8/qJoSOIIKVAQl27Pby00DBAv1zk3j98wK8XsH+zOSE3wf7RcLusbeidP9uzVOccMe2F6d8ilghgN2/VPoLVvGTYh2uEXQRzBz7m3+/xdfbbI4mJN6x3Yer0xbM759b5XH/0hl3EchjiiJ6CXY8+ru/i/06OkdKeUPa7t+/AobNKligf8XnmBEQTNNNsIOgu/eF5//3v6QfvgOndvSUj2CR/nkXOTiCdaCjYM8OMJvgAWSdzcH69PPpUxYs0j/3Mn2FAxiCObS90XyUrdlGN4vMjeaMJFPM9hRFdBaslQgNrwYhWEaSKWZ7iiK6C9YidtOLrQiWkWSK2Z6iiAGCVfehbSMRLCPJFLM9RREjBKurWMePujgPdsncNUB2kg6ZN8vifNbMvc9e+GHYGEMEq6dF+1vxCJaRpH3m28vC24C7D1dlH4YNMkiwOmp0/sf7nQe7ZO4aIDtJ+8yB+4C6iBo32SMME4wBysrcNUB2kvaZr/86987wnNuA/3MDVAHBcpLQH52kQ+b/ru62+gzPEezdVZVPsnggWE4S+qOT9LmA55zlOYJ97nOaiGA5SeiPTtLhFPH+9K7oCFbrs5geCJaThP7oJH0u05d933Rb57OYHgiWk4T+6CRT3OOdoggEy0pCf3SSKWZ7iiIQLCsJ/dFJppjtKYpAsKwk9EcnmWK2pygCwbKS0B+dZIrZnqIIBMtKQn90kilme4oiECwrCf3RSaaY7SmKQLCsJPRHJ5litqcoAsGyktAfnWSK2Z6iCATLSkJ/dJIpZnuKIhAsKwn90UmmmO0pikCwrCT0RyeZYranKALBspLQH51kitmeoggEy0pCf3SSKWZ7iiIQLCsJ/dFJppjtKYpAsKwk9EcnmWK2pygCwbKS0B+dZIrZnqIIBMtKQn90kilme4oiECwrCf3RSaaY7SmKQLCsJPRHJ5litqcoAsGyktAfnWSK2Z6iCATLSkJ/dJIpZnuKIhAsKwn90UmmmO0pikCwrCT0RyeZYranKALBspLQH51kitmeoggEy0pCf3SSKWZ7iiIQLCsJ/dFJppjtKYoYINhyTHagupXZWcIP/nqu0vYVLexXQDPBKm5+5x3wPIr31+ocx87J9rjxHRRLFyzp1WtBHrcsa/v6FtBGsF+7N6fyKuWU7IC13Kck2O8lzRUbItifFRnb17eAJoKVbH8VwYp2wFruExJsWfl7A0YItqz+kLy6eQEtBCva/hqCle2AtVAnI9jBm0pbwwYItsgfU5c3LqC+YIW7t1ywqvN1koIVri9K5mYs3r7SzetbQAPByoJVEKxs+epiBIskczNWn+/yCWtZwCsQrGhzTlCwKldJCrI5CRGs2kRmRSsWrO58IVhyNidh/QHp3J+09QgWjjZOsEQCIVuW6yQs3r6kbMML6CFYYfm9d8Da2tmOYGtFnJ5gCZF7HsGqFDDuCBZ/uPMOWFuLYOEqEEw/h2DWWgQLV9FSsH4XOaoUgGCaExTs6ImW1SKYU0B1wcK7t5VgxTtgbS2ChavoKljq5vUt4BUIVmtrTkew58809WuAYIdPdBcsqYAGggV3bzvBas7XKQr2/Km2fo0Q7PkzydvXt4D6gkXTNxOsdAeshTohwf48d8JfVwm9gbT8ukqVAloIFkvfULCyHbCW5JQEewlfuJRBenzhskoBTQQLfeGypWAn/YXLyL33SBG6/bVIF6zS9q0vdOhbQBvB7gK7d22rhu+AtSSdBAszRRHtBmj89rXbii6bVvPg0g4Ec0CwjCAIZiVBMAsEywiCYFYSBLNAsIwgCGYlQTALBMsIgmBWEgSzQLCMIAhmJUEwCwTLCIJgVpK+gm2W5e0P+QqniO2ynFWsJ6mKQH+Kt+9ue+FnUcvd/sgCtvvbQIEKGgkWSK+TBMZDB7jfgZ+dABFGCXZ7+c17iS5i9+Hq9kuNDuRU4fenePvududFggX6423F7r1+h1gPUmV+vPQySWQ8ZIDrd1eBfegzSrDAznPeocre38NkDlDx9t1efirawkB/vK3YRN6/2gnmpXcF8/zQR/CLOjM2SrDrv86LzmDuNv9zA1Qhc4CKt297UbZ7A/3xDqEfrgJ5mgnmpnfOcM7feMcf9wj25YQFu/vv6m5bcgazeXcVe4stJHeACrfvfrwKBfP742zFJpS/mWBuet+PkkPg/e9gb8rOIY6SdL+K6EyQs/2f+5wmFgxQyfZFrzGsEuiP3oqbj4HfwNoJ5qd3z/Cuy47gwXNkh2GniPcbX3QE2weY+AhWvH2lbx+B/rhXkSJ5Wgnmp/eOYO4xUAc4i13k8Rh5mb5k+x/e43tc5yi4TF+2fRUu0xcV4L3/qyAV5sdP715lf9WX6QNMUQQ3mnOCcKPZSoJgFgiWEQTBrCQIZoFgGUEQzEqCYBYIlhEEwawkCGaBYBlBEMxKgmAWCJYRBMGsJAhmgWAZQRDMSoJgFgiWEQTBrCQIZoFgGUEQzEqCYBYIlhEEwawkCGaBYBlBEMxKgmAWCJYRBMGsJAhmgWAZQRDMSoJgFgiWEQTBrCQIZoFgGUEQzEqCYBYIlhEEwawkCGaBYBlBEMxKgmAWCJYRBMGsJAhmgWAZQRDMSoJgFgiWEQTBrCQIZoFgGUEQzEqCYBYIlhEEwawkCGaBYBlBEMxKgmAWCJYRBMGsJAhmgWAZQRDMSoJgFgiWEQTBrCQIZoFgGUEQzEqCYBYIlhEEwawkCGaBYBlBVOTlmJy05Uv7FIBgDgiWEUQKVqGOP3OdsQ19C0AwBwTLCNJYsKcL0g8/fQtAMAcEywjSVrBF/pi6vHEBCOaAYBlBEMx6EsEsECwjSFPBjl5eur5pAQjmgGAZQRDMeg7BLBAsI0hLwer70bQABHNAsIwgCGY9h2AWCJYRJFGwNJKyDS8AwRwQLCNIlSNY+OFWR7AqBcwnWOg9o2M54QejAUdvXt0CEOzkBAs91I36glWNNryA7oL1u4pYpQAEc0Cw1GhOwOrznVp93wIQzAHBUqM5AYvn+/CJ7oIlFYBgDgiWGs0JWHu+k4vvWwCCOSBYajQnYPl8P/sAe3rtfQtAMAcES43mBKww338mPOcCaN8CEMwBwVKjOQHDdwmcU6/s+wt9C0AwBwRLjVYYsG6Q4QUgmAOCpUYrDFg3yPACEMwBwVKjFQasG2R4AQjmgGCp0QoD1g0yvAAEc0Cw1GiFAesGGV4AgjkgWGq0woB1gwwvAMEcECw1WmHAukGGF4BgDgiWGi0S8Objj7u762X5nBr56fqff2axX3r7ZVnOSgrY6A04DIJgFgiWGi0Q8Oaftz/ubv79tnsvDBFBHtb//DOLh6Xbs3vHhCFeAddnAcURzAHBUqP5AW+/fr8fzN2Hq9vLb4mRn6x//DOLP0u3F/kF3EWOoRMKNvwrv8/LCT8YDTh68wZ/o/mB/WDeHwDuNnkHkDqniF4Ev4DtKZ4iTka7X+JfBict2P1ZqniNX4DegMMgCGaBYJp8wfJPEWsJtvtb+RUSTJ1iHgZBMAsE0+QLVnCRo45gMnugAO8QfBgEwSwQTHPCl+k3+18/sy5y/LlM7x3AEMwDwTTcaA4HQTALBNMgWDgIglkgmAbBwkEQzALBNAgWDjKVYFvnF9ButBqgrf4cXHsiH6QLgGDhIFMJdo93FbUPjQZofxNIfQ6uOdfv9F2oKAgWDjKbYO4thi40FKzGfGezv0Pq3iUNgGDhIJMJdj+Co0vY02qAdudvRvr1cAT7gmDNC5hXsM0Mv4E1G6CH+R56iN4sbz4hWPMCphWs4C59VRoN0P7s7HrsVY46J+EIFg4yl2CTnCE2PIKNPUbf213lKhKChYPMJdjwt/eftBqgDZfpcyL35GULNgvcaNYgWDgIglkgmAbBwkEQzALBNAgWDoJgFgimQbBwEASzQDANgoWDIJgFgmkQLBwEwSwQTINg4SAIZoFgGgQLB0EwCwTTIFg4CIJZIJgGwcJBEMwCwTQIFg6CYBYIpkGwcBAEs0AwDYKFgyCYBYJpECwcBMEsEEyDYOEgCGaBYBoECwdBMAsE01Tvz8v9HxQimAWCaeoLVjXa8AIQzAHBNAgWjoZgFgimQbBwNASzQDANgoWjIZgFgmkQLBwNwSwQTINg4WgIZoFgGgQLR0MwCwTTIFg4GoJZIJgGwcLREMwCwTQIFo6GYBYIpkGwcDQEs0AwDYKFoyGYBYJpECwcDcEsEEyDYOFoCGaBYBoEC0dDMAsE0yBYOBqCWSCYBsHC0RDMAsE09QXjG82vCgTT0B8NgjkwQBr6o0EwBwZIQ380CObAAGnojwbBHBggDf3RIJgDA6ShPxoEc2CANPRHg2AODJCmTX9uPv7Y/2d7URqoqIDNsrz9URgIwRwQTNOkPzf/PAz27nyUYA8F3F5+K4+EYA4IpmnRn9uv3/cHkNvLT4MEeyxg97708HWHYC4Ipml4iri9GHuKeP3X+XJWGgjBHBBM006w3Yerwb+D/Xd1t/1cGAjBHBBM006w7f5jtsN+CXu8ylJ+mQXBHBBM84KvIl7fnx9yBGsNgmlesGD7y/TFBSCYA4Jp6I8GwRwYIA390SCYAwOkoT8aBHNggDT0R4NgDgyQhv5oEMyBAdLQHw2COTBAGvqjQTAHBkhDfzQI5sAAaeiPBsEcGCAN/dEgmAMDpKE/GgRzYIA09EeDYA4MkIb+aBDMgQHS0B8NgjkwQBr6o0EwBwZIQ380CObAAGnoj2ZFMPiN2TX4Df3RmIIBQG0QDKAhCAbQEAQDaAiCATQEwQAagmAADUEwgIYgGEBDEAygIf8H28HP13DBaUwAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="pedigree-subsets" class="section level2" number="1.5">
<h2><span class="header-section-number">1.5</span> Pedigree subsets</h2>
<p>Many situations call for selecting all pedigree members sharing some
property, e.g., all females, or all descendants of some person. Several
utilities in <strong>pedtools</strong> exist to help with such tasks.
Generally these come in two flavours: 1) members with certain
<em>global</em> property, and 2) members with a certain relationship to
a given individual.</p>
<p><strong>Pedigree members with a certain property</strong><br />
Each of the following functions returns a vector specifying the members
with the given property.</p>
<ul>
<li><code>founders()</code></li>
<li><code>nonfounders()</code></li>
<li><code>leaves()</code></li>
<li><code>males()</code></li>
<li><code>females()</code></li>
<li><code>typedMembers()</code></li>
<li><code>untypedMembers()</code></li>
</ul>
<p>By default, the output of these functions is a character vector
containing ID labels. However, adding the option
<code>internal = TRUE</code> will give you an integer vector instead,
reporting the internal indices of the members. This is frequently used
in the source code of <code>pedtools</code>, but is usually not intended
for end users of the package.</p>
<p><strong>Relatives of a given individual</strong><br />
The functions below take as input a <code>ped</code> object and the
label of a single member. They return a vector of all members with the
given relation to that individual.</p>
<ul>
<li><code>father()</code></li>
<li><code>mother()</code></li>
<li><code>parents()</code></li>
<li><code>grandparents()</code></li>
<li><code>children()</code></li>
<li><code>spouses()</code></li>
<li><code>siblings()</code></li>
<li><code>cousins()</code></li>
<li><code>nephews_nieces()</code></li>
<li><code>ancestors()</code></li>
<li><code>descendants()</code></li>
<li><code>unrelated()</code></li>
</ul>
</div>
</div>
<div id="markers" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Markers</h1>
<p>The other main theme of the <code>pedtools</code> package (pedigrees
being the first) are marker genotypes.</p>
<div id="creating-marker-objects" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Creating marker
objects</h2>
<p>Marker objects created with the <code>marker()</code> function. For
example, the following command makes an empty marker associated with the
<code>trio</code> pedigree:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">marker</span>(trio)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id  &lt;NA&gt;</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   -/-</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   -/-</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl   -/-</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; * * * * * *  </span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Position: NA </span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Mutation: none </span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Frequencies:</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    1   2</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  0.5 0.5</span></span></code></pre></div>
<p>As shown in the output, the marker is indeed empty: All pedigree
members have missing genotypes, and there is no assigned name or
position. By default, markers are diallelic, with alleles 1 and 2, with
equal frequencies. For a more interesting example, let us make a SNP
named “snp1”, with alleles “A” and “B”. The father is homozygous “A/A”,
while the mother is heterozygous. We store it in a variable
<code>m1</code> for later use.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">=</span> <span class="fu">marker</span>(trio, <span class="at">fa =</span> <span class="st">&quot;A/A&quot;</span>, <span class="at">mo =</span> <span class="st">&quot;A/B&quot;</span>, <span class="at">name =</span> <span class="st">&quot;snp1&quot;</span>)</span></code></pre></div>
<p>This illustrates several points. Firstly, individual genotypes may be
specified using the ID labels. The different alleles occurring in the
genotypes is interpreted as the complete set of alleles for the marker.
Finally, these are assigned equal frequencies. Of course, this behaviour
can be overridden, by declaring alleles frequencies explicitly:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">marker</span>(trio, <span class="at">fa =</span> <span class="st">&quot;A/A&quot;</span>, <span class="at">mo =</span> <span class="st">&quot;A/B&quot;</span>, <span class="at">afreq =</span> <span class="fu">c</span>(<span class="at">A =</span> .<span class="dv">2</span>, <span class="at">B =</span> .<span class="dv">3</span>, <span class="at">C =</span> .<span class="dv">5</span>))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id  &lt;NA&gt;</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   A/A</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   A/B</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl   -/-</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; * * * * * *  </span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Position: NA </span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Mutation: none </span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Frequencies:</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    A   B   C</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  0.2 0.3 0.5</span></span></code></pre></div>
<p>The markers chromosome can be declared using the <code>chrom</code>
argument, and similarly its position by <code>posMb</code> (megabases).
Markers with unknown chromosome are treated as autosomal. To define an
X-linked marker, put <code>chrom = &quot;X&quot;</code>. the fact that males are
hemizygous on X (i.e. they have only one allele) is reflected in the
printout of such markers:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>m2 <span class="ot">=</span> <span class="fu">marker</span>(trio, <span class="at">fa =</span> <span class="st">&quot;A/A&quot;</span>, <span class="at">mo =</span> <span class="st">&quot;A/B&quot;</span>, <span class="at">chrom =</span> <span class="st">&quot;X&quot;</span>, <span class="at">name =</span> <span class="st">&quot;snpX&quot;</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>m2</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id  snpX</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa     A</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   A/B</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl   -/-</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; * * * * * *  </span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Position: chr = X, Mb = NA </span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Mutation: none </span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Frequencies:</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    A   B</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  0.5 0.5</span></span></code></pre></div>
<p>A side note: It may come as a surprise that you don’t need quotes
around the ID labels (which are characters!) in the above commands. This
is because <code>marker()</code> uses <em>non-standard evaluation
(NSE)</em>, a peculiarity of the R language which often leads to less
typing and more readable code.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> Unfortunately, this doesn’t work with
numerical ID labels. Thus to assign a genotype to someone labelled “1”
you need quotes, as in <code>marker(trio, &quot;1&quot; = &quot;A/A&quot;)</code>.</p>
</div>
<div id="plotting-pedigrees-with-marker-data" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Plotting pedigrees
with marker data</h2>
<p>Including marker data in a pedigree plot is straightforward:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(trio, <span class="at">marker =</span> m1)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAXVBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6Ojo6OpA6kNtmAABmADpmZjpmtrZmtv+QOgCQOjqQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///87oUNWAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAD7ElEQVR4nO2cD3ebIBTFSZZsS7bZNWyrM2m//8ecaIx/IipP4MrOvaftKTW+y094CPaI+khcCl2BtSIAWgRAiwBoEQAtAqBFALQIgBYB0PovAdSUIlZtkeUogOPnQ6hb70mEjQKoyeLcITzAk4u9ETYJ4FKpLQKMe1icNwjgUFPLn7EATr2dAAFkdxg9sjkAHzctAqwRAeZiBFdogIljPj6/uS5EgLljmwPY2o2MAHMHCCCtjvMpoPWA5M69uK/b/o4GGF3CJwXwfBruqYR08tev8USUrQK0Z0omeMLJ4tL4S09dZBh6fg9dP6QQnwDo+ARAxycAOj4B0PEnDXKlTiHje9GEwe28uvpogCxofD+yG1yPSu3fPnQ5IVwBIgG4Hn+cy95rKmCcCzXVledaQB/KXNhdBNWYjW/V9Vga5ubq5eV3UVLczgeBgQG4fbuYgPImkAGcHj+y9xdz9QvrNVyQA8WaPiQDyNof9dWzX8M5gHIo3f+O3gJPAPYBZQagjSUUugWKMouqNJJqNcDKHDBhbucVN+TVACtHIZMDu4uWj6PrAeT3AS/ibBRtQAC0AQHQBjgA3Yy9169/u0Vf8SfVNdfVM2rnO/H183cziyhVHHpFR4kAeub6k7mA1jWJzSB/zEF11is6SgTQM68BrAt0i8H7y6H8qn77eekWXSUB6Js3AJbrZzEws7+61aoUaIuukgD0ze9dyHb5LAbmrLrVqhRoi66SAPTN6yRWFcZig2pB+qFNJumsV3SVAGBgXrdAoSxNMG6Q3//BltUp0BadJQAYmNcA1ss3anBPIbOMML2wU3SWO8DAvAHIXQCaBVyZSaYXdorO1XEHGJg/WsCSBKMGzYfL3tj2wqZzuskdYGD+yAGHYbR9BKB3Xy7donsaOwMMzJtRSLKo9yLORtEGBEAbEABtQAC0AQHQBgRAGxAAbUAAtAEB0AYEQBsQAG1AALQBAdAGBEAbpA0ge3PL0SNgaPX4JaRLsMDKXvLqEytuNKNgYQM1QsSoCQFYgsb0ChMzqlmImIkATISM6xYgJAFihJyMGNnOf0QCoCMSIELE5EchAnhz8x8zkblQ+gDJrwdsURMHSGpNPBI2sacSw7jpPRfq7yyV4pO5KrhwXysnj4Cxo4gAaBEALQKgRQC0CDCpxyucXjZbGlWkFiCAVYEAtFK7192l3mPpVe3/JAagq62l7gDn6u3apADqt/F0A3BKrgtVuxKZtzprgCw5gDx1gORboM6BPF2AwSiUIIC5D+x/7d/SBTAqZJuZOCpcDnjZ7XNeYVqgEG7iIBDXA2gRAC0CoEUAtAiAFgHQIgBaBECLAGgRAK1/aAx9VyMfF5kAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>The appearance of the genotypes can be tweaked in various ways, as
documented in <code>?plot.ped</code>. Here’s an example:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(trio, <span class="at">marker =</span> m1, <span class="at">showEmpty =</span> T, <span class="at">missing =</span> <span class="st">&quot;?&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAVFBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6kNtmAABmADpmZjpmtrZmtv+QOgCQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///9DYrSiAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAEFklEQVR4nO2ci3KbMBBFZddu3NaJaRuKH///n0U8hOywwlqxXOTZO01GicNeHUkLkgqYW+Yy6AqkSgHQUgC0FAAtBUBLAdBSALQUAC0FQOslAUxAi9btGdNRgLg/l5Ff75DrWgHujQKdsE6ArxUmjVcJEFOpNQKMmxDDKB8Aiuv5ABMfzaeY0bJGgKh8VYD5FemuALNLGGAByQLQH810wNqGkAJMfbQ+gMAhuQNALmSc8+6q5kIKMPXrNQI839TEr+EAY6sv1JKSO3l6PA62qGfP/u46AbitkjB97ed14a25UQDWVPH5+BFHT1tKT/DFFxAKAI6vAOj4CoCOrwDo+C8NUBrzJhl/HtEGl0N69cEAR9H4M4k0OO+N2X7eino6mALCAjjvfx3q8WurYL0rExjMEz1Q7Opc2Jw4tZiKH9B5X1uWtv3K+quqKS6HXbSBBbj8ONlwCV3ABHhz347Xd9v6FdWK0zlQJY0hJsBx+Na2H9mKEwD1qXT7B9ADXwDIU0oYYIjEFbgHqjqHmiRiKx0gKQdskMsh5YKcDpB0FrI5sDkVCefRGQC414F5pLNRtIECoA0UAG0ABCiGc2/Bv5SxAXrPotmhpq9E9ILm+087j3goRosL4DyLb/9uoTUJaVAOs9AyYULKBXCeLQC9QKcMru+7+t9jMV5MgMGzB6BakDKws7+u37xivJgAg2c3hMgGpAzscV2/ecV4MQEGzzaJTYMRYdAsSG+FzSOvyBAPwPNse6AykdPpsvs/veNdkSEegOfZAtANOG7QJZBdRnhFjlgAvmcHUMYB9Au4Oo+8IqcuLADfs+8BKgnGDfo/r8eiV+TUhQXge/Y5EHUaHbYAimHs89KYA+DZbz/ZU4mZpLNRtIECoA0UAG2gAGgDBUAbKADUgHmnY5yHYGjjCpIuYoEN/dOsPkvFXcxILOxyTlJRF7SSCZoPABVzwcYSirmomUTMTAACIZd1EwipAOiQCoAOqQDokHkAZH8he12AbCZz2U+n81/QjEbNakk5EjazRf3jPkp+2yp321lZbmw1wRfYWxTffJWWAqClAGgpAFoKgJYCBOXuGJ7n2d4xLdQDCkBKCKAwZvOxObWP9H6Y7d/MAIrmOeYO4NDcyJ0VQHvjZ9EDvGU3hJqHYO0NxC3AMTuAMneA7HugzYEyX4CHs1CGAPY6sP29/cwXwKpiPz0XI7kcmOflMpOS6YGK/9RQrHQ9gJYCoCUG4N5mEXytRbqkAOwzbOf9zisISQjAvdFl4tUu6RICcO/UmXi5TrpEk9g1fH490MgN/QxzwOq8d890ks9jp0sO4PreVdsVRCQH4BJXMINvkgBLZPBNpxIBaQ88JwVASwHQUgC0FAAtBUBLAdBSALQUAC0FQEsB0FIAtP4D7BXWC2XGYRIAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
</div>
<div id="markers-attached-to-ped-objects" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Markers attached to
<code>ped</code> objects</h2>
<p>In most applications it is useful to <em>attach</em> markers to their
<code>ped</code> object. In particular for bigger projects with many
markers, this makes it easier to manipulate the dataset as a unit.</p>
<p>To attach a marker object <code>m</code> (which could be a list of
several markers) to a pedigree <code>x</code>, there are two main
options:</p>
<ul>
<li><code>setMarkers(x, m)</code></li>
<li><code>addMarkers(x, m)</code></li>
</ul>
<p>The difference between these is that <code>setMarkers()</code>
replaces all existing markers, while <code>addMarkers()</code> appends
<code>m</code> to the existing ones. In our <code>trio</code> example
the two are equivalent since there are no existing markers.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>trio <span class="ot">=</span> <span class="fu">setMarkers</span>(trio, <span class="fu">list</span>(m1, m2))</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>trio</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id fid mid sex snp1 snpX</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   *   *   1  A/A    A</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   *   *   2  A/B  A/B</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl  fa  mo   2  -/-  -/-</span></span></code></pre></div>
<p>There is a handy shortcut, <code>addMarker()</code> (without the
‘s’), allowing you to create and attach a single marker in one go. The
command <code>addMarker(x, ...)</code> is essentially equivalent to
<code>setMarkers(x, marker(x, ...))</code>. It is also well adapted to
piping, as in this example:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nuclearPed</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addMarker</span>(<span class="at">name =</span> <span class="st">&quot;myMarker&quot;</span>, <span class="at">alleles =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>)) <span class="sc">|&gt;</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setGenotype</span>(<span class="at">marker =</span> <span class="dv">1</span>, <span class="at">id =</span> <span class="dv">3</span>, <span class="at">geno =</span> <span class="st">&quot;a/c&quot;</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  id fid mid sex myMarker</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   1   *   *   1      -/-</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   2   *   *   2      -/-</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   3   1   2   1      a/c</span></span></code></pre></div>
<p><strong>Selecting and removing attached markers</strong><br />
Four closely related functions functions are useful for manipulating
markers attached to a pedigree:</p>
<ul>
<li><code>selectMarkers()</code>, returns a <code>ped</code> object
where only the indicated markers are retained</li>
<li><code>removeMarkers()</code>, returns a <code>ped</code> object
where the indicated markers are removed</li>
<li><code>getMarkers()</code>, returns a list of the indicated
markers</li>
<li><code>whichMarkers()</code>, returns the indices of the indicated
markers</li>
</ul>
<p>All of these have exactly the same arguments, described in more
detail in <code>?marker_select</code>. Let us do a couple of examples
here. Recall that by now, our <code>trio</code> has two attached
markers; the first is called “snp1”, and the other is on the X
chromosome.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">whichMarkers</span>(trio, <span class="at">chrom =</span> <span class="st">&quot;X&quot;</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 2</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">selectMarkers</span>(trio, <span class="at">markers =</span> <span class="st">&quot;snp1&quot;</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id fid mid sex snp1</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   *   *   1  A/A</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   *   *   2  A/B</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl  fa  mo   2  -/-</span></span></code></pre></div>
</div>
<div id="accessing-and-modifying-individual-markers" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Accessing and
modifying individual markers</h2>
<p>Internally, a marker object is stored as a matrix with two columns
(one for each allele) and one row for each pedigree member. The matrix
is numeric (for computational convenience) while the allele labels and
other meta information are added as <em>attributes</em>. The most
important of these are:</p>
<ul>
<li><code>alleles</code> : The allele labels, stored as a character
vector.</li>
<li><code>afreq</code> : The allele frequencies, in the same order as
the alleles. An error is issued if the frequencies do not sum to 1 after
rounding to 3 decimals.</li>
<li><code>name</code> : The marker name, which can be any character
string not consisting solely of digits.</li>
<li><code>chrom</code> : The chromosome name. This can be given as an
integer, but is always converted to character. The special values “23”
and “X” are recognized as the human X chromosome, which affects the way
genotypes are printed.</li>
<li><code>posMb</code> : Chromosomal position given in megabases.</li>
</ul>
<p>In addition to those listed above, there are two more attributes:
<code>pedmembers</code> and <code>sex</code>. They store the ID labels
and genders of the pedigree associated with the marker, and are only
used to empower the printing method of marker objects.</p>
<p><strong>Marker accessor functions</strong><br />
For each marker attribute listed above, there is a corresponding
function with the same name for retrieving its content. These functions
take as input either a <code>marker</code> object, or a <code>ped</code>
object together with the name (or index) of an attached marker. This may
sound a bit confusing, but a few examples will make it clear!</p>
<p>Recall that our marker “snp1” exists in two copies: One is stored in
the variable <code>m1</code>, while the other is attached to
<code>trio</code>. In both cases we can extract the allele frequencies
with the function <code>afreq()</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">afreq</span>(m1)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   A   B </span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0.5 0.5</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">afreq</span>(trio, <span class="at">marker =</span> <span class="st">&quot;snp1&quot;</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   A   B </span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0.5 0.5</span></span></code></pre></div>
<p>We can also <em>modify</em> the frequencies using this syntax. To
avoid confusion about the allele order, the frequencies must be named
with the allele labels (just as in the output of <code>afreq()</code>
above).</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">afreq</span>(trio, <span class="at">marker =</span> <span class="st">&quot;snp1&quot;</span>) <span class="ot">=</span> <span class="fu">c</span>(<span class="at">A =</span> <span class="fl">0.9</span>, <span class="at">B =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<p>In addition to the functions getting and setting marker attributes,
there is one more important marker accessor, namely
<code>genotype()</code>. This returns the genotype of a specified
individual, and can also be used to modify genotypes. As the others, it
can be applied to marker objects directly, or to pedigrees with attached
markers. Here we show a few examples of the latter type:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Girl is not genotyped</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">genotype</span>(trio, <span class="st">&quot;snpX&quot;</span>, <span class="at">id =</span> <span class="st">&quot;girl&quot;</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] NA NA</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Set genotype</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">genotype</span>(trio, <span class="st">&quot;snpX&quot;</span>, <span class="at">id =</span> <span class="st">&quot;girl&quot;</span>) <span class="ot">=</span> <span class="st">&quot;A/A&quot;</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the result</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>trio</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    id fid mid sex snp1 snpX</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    fa   *   *   1  A/A    A</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    mo   *   *   2  A/B  A/B</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  girl  fa  mo   2  -/-  A/A</span></span></code></pre></div>
</div>
<div id="gettingsettingmodifying-many-markers-simultaneously" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span>
Getting/setting/modifying many markers simultaneously</h2>
Several functions in <code>pedtools</code> are indented for modifying
many (or all) markers at the same time. Their purpose and typical use
cases are summarised in the table below. The argument <code>x</code>
always denotes a <code>ped</code> object.
<table>
<thead>
<tr>
<th style="text-align:left;">
Use …
</th>
<th style="text-align:left;">
When you want to …
</th>
<th style="text-align:left;">
For example to …
</th>
</tr>
</thead>
<tbody>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Get</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>getAlleles(x)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
extract all alleles as a matrix.
</td>
<td style="text-align:left;">
do summary stats on the marker alleles
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>getFreqDatabase(x)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
extract allele frequencies as a data.frame in <em>allelic ladder</em>
format.
</td>
<td style="text-align:left;">
transfer to other objects, or write the database to a file
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>getMarkers(x)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
extract list of marker objects. Each marker is a <code>N * 2</code>
allele matrix (<code>N = pedsize(x)</code>) with locus annotations as
attributes
</td>
<td style="text-align:left;">
do computations
</td>
</tr>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Set</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>setAlleles(x, ...)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
replace the genotypes of <code>x</code> without changing the locus
attributes.
</td>
<td style="text-align:left;">
erase all genotypes
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>setFreqDatabase(x, db)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
replace all allele frequencies without changing the genotype data. The
input is a data.frame in <em>allelic ladder</em> format. Conceptually
equivalent to
<code>setMarkers(x, alleleMatrix = getAlleles(x), locusAnnotations = db)</code>.
</td>
<td style="text-align:left;">
change the frequency database
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>setMarkers(x, ...)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
attach marker objects with or without genotype data. Locus attributes
are indicated as a list; genotypes as a matrix or data.frame.
</td>
<td style="text-align:left;">
prepare joint manipulation of a pedigree and marker data
</td>
</tr>
<tr grouplength="2">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Convert</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>as.data.frame(x)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
convert <code>x</code> to a data.frame, with pedigree columns in
standard format followed by genotype columns. One column per marker,
with genotype format <code>a/b</code> and missing alleles indicated as
<code>-</code>.
</td>
<td style="text-align:left;">
pretty-print ped objects
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>as.matrix(x)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
convert <code>x</code> to a <em>numerical</em> matrix, with additional
info attached as attributes.
</td>
<td style="text-align:left;">
modify a pedigree with marker data
</td>
</tr>
<tr groupLength="1">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Transfer</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 4.5cm; ">
<code>transferMarkers(from, to)</code>
</td>
<td style="text-align:left;width: 8.5cm; ">
transfer genotypes and attributes between pedigree objects (or lists of
such).
</td>
<td style="text-align:left;">
transfer simulated marker data
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>There is an important exception to this: Certain
algorithms in pedigree analysis work “top-down”, in the sense that
parents must be treated before their children. For this reason, many
implementations require, for simplicity, that the individuals are stored
in this fashion, i.e. that parents always precede their children.
<code>pedtools</code> offers a special reordering function to ensure
this, <code>parents_before_children()</code>, which you will find
utilised in the source code of packages like <code>ribd</code> and
<code>ibdsim2</code>.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>You may have come across NSE before, for instance when
using <code>subset()</code> on a data.frame. To learn more about NSE, I
recommend this book chapter by Hadley Wickham:<br />
<a href="http://adv-r.had.co.nz/Computing-on-the-language.html" class="uri">http://adv-r.had.co.nz/Computing-on-the-language.html</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
