<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Magnus Dehli Vigeland" />

<meta name="date" content="2020-06-18" />

<title>Introduction to the pedtools package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>$(document).ready(function(){
    if (typeof $('[data-toggle="tooltip"]').tooltip === 'function') {
        $('[data-toggle="tooltip"]').tooltip();
    }
    if ($('[data-toggle="popover"]').popover === 'function') {
        $('[data-toggle="popover"]').popover();
    }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to the pedtools package</h1>
<h4 class="author">Magnus Dehli Vigeland</h4>
<h4 class="date">2020-06-18</h4>


<div id="TOC">
<ul>
<li><a href="#pedigrees"><span class="toc-section-number">1</span> Pedigrees</a>
<ul>
<li><a href="#the-ped-class"><span class="toc-section-number">1.1</span> The <code>ped</code> class</a></li>
<li><a href="#basic-pedigree-plots"><span class="toc-section-number">1.2</span> Basic pedigree plots</a></li>
<li><a href="#builtin"><span class="toc-section-number">1.3</span> Built-in pedigree structures</a></li>
<li><a href="#examples-of-pedigree-construction"><span class="toc-section-number">1.4</span> Examples of pedigree construction</a></li>
<li><a href="#pedigree-subsets"><span class="toc-section-number">1.5</span> Pedigree subsets</a></li>
</ul></li>
<li><a href="#markers"><span class="toc-section-number">2</span> Markers</a>
<ul>
<li><a href="#creating-marker-objects"><span class="toc-section-number">2.1</span> Creating marker objects</a></li>
<li><a href="#plotting-pedigrees-with-marker-data"><span class="toc-section-number">2.2</span> Plotting pedigrees with marker data</a></li>
<li><a href="#markers-attached-to-ped-objects"><span class="toc-section-number">2.3</span> Markers attached to <code>ped</code> objects</a></li>
<li><a href="#accessing-and-modifying-individual-markers"><span class="toc-section-number">2.4</span> Accessing and modifying individual markers</a></li>
<li><a href="#gettingsettingmodifying-many-markers-simultaneously"><span class="toc-section-number">2.5</span> Getting/setting/modifying many markers simultaneously</a></li>
</ul></li>
</ul>
</div>

<p>The purpose of this vignette is to show how to work with pedigrees and marker data in <code>pedtools</code>.</p>
<p>The following command installs the current CRAN version of the package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">install.packages</span>(<span class="st">&quot;pedtools&quot;</span>)</span></code></pre></div>
<p>Alternatively, you may want the latest development version from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># install.packages(&quot;devtools&quot;) # install devtools if needed</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;magnusdv/pedtools&quot;</span>)</span></code></pre></div>
<p>Now you should be able to load <code>pedtools</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">library</span>(pedtools)</span></code></pre></div>
<div id="pedigrees" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Pedigrees</h1>
<p>In <code>pedtools</code> and all related packages, pedigrees are stored as <code>ped</code> objects. We start by explaining briefly what these objects look like, and their basic constructor. If you are reading this vignette simply to learn how to create a particular pedigree, you may want to skip ahead to section 1.3 where we describe practical shortcuts to common pedigree structures.</p>
<div id="the-ped-class" class="section level2" number="1.1">
<h2 number="1.1"><span class="header-section-number">1.1</span> The <code>ped</code> class</h2>
<p><strong>The <code>ped</code> constructor function</strong><br />
The most direct way to create a pedigree in <code>pedtools</code> is with the <code>ped()</code> constructor. This takes as input 4 vectors of equal length:</p>
<ul>
<li><code>id</code> : individual ID labels (numeric or character)</li>
<li><code>fid</code> : id of the fathers (0 if not included)</li>
<li><code>mid</code> : id of the mothers (0 if not included)</li>
<li><code>sex</code> : gender codes, with entries 0 (unknown), 1 (male) or 2 (female)</li>
</ul>
<p>In other words, the j’th pedigree member has label <code>id[j]</code>, father <code>fid[j]</code>, mother <code>mid[j]</code>, and gender given by <code>sex[j]</code>.</p>
<p>For example, the following creates a family <em>trio</em>, i.e. father, mother and child:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">ped</span>(<span class="dt">id =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">fid =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">mid =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>), <span class="dt">sex =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt;  id fid mid sex</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt;   1   *   *   1</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt;   2   *   *   2</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt;   3   1   2   2</span></span></code></pre></div>
<p>In this example the child (<code>id=3</code>) is female, since the associated entry in <code>sex</code> is 2. Note that missing parents are printed as <code>*</code>. Individuals without parents are called <em>founders</em> of the pedigree, while the <em>nonfounders</em> have both parents specified. It is not allowed to have exactly one parent.</p>
<p>Instead of numerical labels as above, we could have used character strings. Let us create the trio again, with more informative labels, and store it in a variable named <code>trio</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>trio =<span class="st"> </span><span class="kw">ped</span>(<span class="dt">id =</span> <span class="kw">c</span>(<span class="st">&quot;fa&quot;</span>, <span class="st">&quot;mo&quot;</span>, <span class="st">&quot;girl&quot;</span>), <span class="dt">fid =</span> <span class="kw">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;fa&quot;</span>), <span class="dt">mid =</span> <span class="kw">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;mo&quot;</span>), <span class="dt">sex =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb5-2"><a href="#cb5-2"></a>trio</span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">#&gt;    id fid mid sex</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">#&gt;    fa   *   *   1</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">#&gt;    mo   *   *   2</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">#&gt;  girl  fa  mo   2</span></span></code></pre></div>
<p>The special strings <code>&quot;0&quot;</code>, <code>&quot;&quot;</code> and <code>NA</code> are all interpreted as a missing parent.</p>
<p><strong>The internal structure of <code>ped</code> objects</strong><br />
From the way it is printed, the object <code>trio</code> appears to be a data frame, but this is not exactly true. Rather it is an object of class <code>ped</code>, which is basically a list. We can see the actual content of <code>trio</code> by unclassing it:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">unclass</span>(trio)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co">#&gt; $ID</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt; [1] &quot;fa&quot;   &quot;mo&quot;   &quot;girl&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; </span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; $FIDX</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">#&gt; [1] 0 0 1</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">#&gt; </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">#&gt; $MIDX</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">#&gt; [1] 0 0 2</span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">#&gt; </span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">#&gt; $SEX</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">#&gt; [1] 1 2 2</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">#&gt; </span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">#&gt; $FAMID</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co">#&gt; [1] &quot;&quot;</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co">#&gt; </span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="co">#&gt; $UNBROKEN_LOOPS</span></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb6-19"><a href="#cb6-19"></a><span class="co">#&gt; </span></span>
<span id="cb6-20"><a href="#cb6-20"></a><span class="co">#&gt; $LOOP_BREAKERS</span></span>
<span id="cb6-21"><a href="#cb6-21"></a><span class="co">#&gt; NULL</span></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="co">#&gt; </span></span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="co">#&gt; $FOUNDER_INBREEDING</span></span>
<span id="cb6-24"><a href="#cb6-24"></a><span class="co">#&gt; NULL</span></span>
<span id="cb6-25"><a href="#cb6-25"></a><span class="co">#&gt; </span></span>
<span id="cb6-26"><a href="#cb6-26"></a><span class="co">#&gt; $MARKERS</span></span>
<span id="cb6-27"><a href="#cb6-27"></a><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>In most cases it is not recommended for regular users to interact directly with the internal slots of a <code>ped</code>, since this can have unfortunate consequences unless you know exactly what you are doing. Instead, one should use accessor functions like <code>labels()</code>, <code>getMarkers()</code> and <code>founderInbreeding()</code>. The most important accessors are described within this vignette, while others are documented in the help page <code>?ped_utils</code>.</p>
</div>
<div id="basic-pedigree-plots" class="section level2" number="1.2">
<h2 number="1.2"><span class="header-section-number">1.2</span> Basic pedigree plots</h2>
<p>To plot a pedigree, simply use <code>plot()</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">plot</span>(trio)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAATlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6kNtmAABmADpmtrZmtv+QOgCQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///8NKk6TAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADAUlEQVR4nO2cC1PqMBSEQxWqKPVRC/r//6gNvRcVm6RNcrI5zu6MjlDYzde8O2PMh3IZdAFSRQC0CIAWAdAiAFoEQIsAaBEALQKg9ScBjE8Fi7YochZg5ecl9L3cXoRKAYz3ZegSHuBXirsSqgRYU6gaAeYzHMkVAqwoqeNtLMCq1q4KYP4KAXIrx6RFgBQRIOQhLmkAz7Ucn6+uCREgdK06gNomMgKELtQHsGrrogkAtB+ImbkXt3XX+2iAuS08bFMfuXa6/hruqUTs4u97ib0LpFoBLsWOerQYuVhc6r/8ywvypNf30P2DBn8CoP0JgPYnANqfAGh/b0BvzE7SP4s8Aac2ufhogL2ofx65A45bY5rXj25cESaAxAAct/ft2HptAWzyYHxNOVQD3e3YFzaHiGIE/Z06bsfA3t69fvwZRopTexsRYAFOdwdrGF8FcQC7y6/9+4O9+4PzHi7oA0NKG4oD2H/9mu6e+x6GAMahtHkuXgO/ANwDSgDgyytS6BoYxl507kaxSgZI7APW5tQmTMjJAImjkO0Dm0MXP46mA8TPA1nE1Sg6gADoAAKgAwiADiAAOoAA6AACoAMIgA4gADqAAOgAAqADCIAOIAA6gADoAAKgAwiADiAAOoAA6AACoAMIgA4gADqAAOgAAqADCIAOIAA6gADoAN0Acf9+uTJD0Npc/pBMETM27ldZc0r5FgsSsxWqhIKuigAcpiWzZDyLhkl4KgHwWJZNE7AkQAlLr2PhuPyOBEA7EqCAo/pRiADZ0vJ7KlkL6QdQvx/QvyPTvyeesVX2VOLaV99zoZ9HrGl8Mnc2jzycblWGoHcREQAtAqBFALQIgBYBvLqca5TlxLRZFaoBAjglBNAZs3ncHKaD0h5N86IMoDufD/cPoL1501YD05Fe3X+AnbomdD5azB7sNQHs1QH02gHU18DUB3q9AFejkEIAOw80T82rXgCrqSdIS64PZDmyNyyZGrCnC6YcdLtC3A+gRQC0CIAWAdAiAFoEQIsAaKkH+AQ7KM57aNZRYQAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
<p>Under the hood, <code>pedtools::plot()</code> is an elaborate wrapper of the excellent plotting functionality of the <code>kinship2</code> package. Most of the possibilities provided by kinship2 are available from pedtools, and several features are added. An overview can be found in the documentation <code>?plot.ped</code>, but a quick example should get you started:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">plot</span>(trio, <span class="dt">deceased =</span> <span class="st">&quot;fa&quot;</span>, <span class="dt">starred =</span> <span class="st">&quot;mo&quot;</span>, <span class="dt">shaded =</span> <span class="st">&quot;girl&quot;</span>, </span>
<span id="cb8-2"><a href="#cb8-2"></a>     <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="dt">title =</span> <span class="st">&quot;Trio 1&quot;</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAjVBMVEUAAAAAADoAAGYAAP8AOjoAOmYAOpAAZrYA/wA6AAA6AGY6ZrY6kJA6kNtmAABmADpmOgBmOjpmkLZmkNtmtrZmtv+QOgCQZgCQkGaQtpCQ2/+2ZgC2Zjq2tma225C229u22/+2/7a2///bkDrbkGbbtpDb29vb/9vb////AAD/tmb/25D//7b//9v///8XqFVOAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAEYUlEQVR4nO2cDXPTMAyG1QKFASvjo2M0wGBhMNI2///nEddpujW2EidSVHF675ZukedXjz+SxpczlMoF0gmMlQFIywCkZQDSogHIoNHLw7ndNSweYsVjkXQJAOy+ggqAmO6v4PwAnKoW98nnMPtyBfPbugfu3wM8/9AknAPMLs4dwGnxxwP4v44Z5/DiNj470sUDsPhR/vZzYLNv7erUZV3s15sHZHqniwdgVR4mcQbzu7I8fDQlzxxgti7rNA+51ueakmoAtkt/svC90pRUA6C+B9TPgfZVqFQG0LoPlNoA/J3440nJcwQQkgFIawDAeTGnZ3Ne+aenc2b5B/N5hWjS/P86dZQJAiDFkRixDrl3IKQBAAZHqsdpo92QBABo71DqNOM4QgoAIDFSBdKlAAAkRqpgsjGC/gCAxEgVSTVyujcAIDFa8QAAEqNVdLSHAz0BAIkRK37JDEb6AQASoxYHACAxamH33VCsDwAgMXIxAAASoxc9wNMSwC5qAEBiPZT89Zu6B07j2gBaYWUA7agugECQHYDyRpb4uBmUJEDQWxAg9ctc2JofgOrrdMRZDUDMeAIAkkfKqO8UAAQP9XHbSQBGL6sgrhMBjFvYwkynAmjS7lodDRgAtrg7HUDpF3fTDWjXn9lXs1sGxI7TAyir3wCk6zcA6foNQLr+/xvg6WtK9PUnKY+8EowYbJej0ycDKGY3b5dBBBRgFQ/2FVUP7K4fvXTXz2BzAe5NPfde9wiQZIDNxadlNXSdu7MtwI/jYvE5PIa6eiCr/i1/9MpkqgYAVG65a7q8+ikqiq0bO9lq8zrYBR0A23drV+fwLhgAcNkcVrvrfetjDdhjDhRjxtAAgNXx4JsObcAugOpSOv8+bQ+0ANCrSQfAsbqBku6Bwl28CrEhNH4OuJq2yxE35HEAx6vQEIPDHJits+HX0ZEAzX2AziBR9m1U2sAApA0MQNrAAKQNDEDawACkDQxA2sAApA0MQNrAAKQNDEDawACkDQxA2sAApA0MQNrAAKQNDEDawACkDQxA2sAApA0MQNrAAKQN4Bm3AV/Vz/baf/CZ8AH4rOsDIwMXwJP8m78ZxARwmr82gHb+bAgsAKH8myOxeACeHlhHEQdAtP05CBgAkPHDQMAEEMlfDUCs/bUARMePCgAkfw4CBgBs/qoAaA6B/LUARNtfA0CT5iT5swDEx78SgNPj8VdFAOH8dQAcx8oE+XM+D7THjx6A45E7f65HynKq/Nke6ttTmempnm1ZpTnoXFbZZzxF/sxLi2W9oMKWPvfibr08yin25W9uGYC0DEBaBiAtA5CWAaBqtrQg2SknqIl6wACiYgLIAGY3s7XfIOcG5j+VAWT7rYFqgOXiQVsP+N1csgPApbohtN9Sxu3p4gFW6gBy7QDqe8DPgVwvwMlVSCGAuw/Mv83v9AI4FZG9+WjFNwdItmrsFk8PuI2lxmxwmCB7HpCWAUjLAKRlANIyAGkZgLTUA/wDE+A2RMANi4sAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>See Section 2.2 for how to add, and control the appearance of, marker genotypes to pedigree plots.</p>
</div>
<div id="builtin" class="section level2" number="1.3">
<h2 number="1.3"><span class="header-section-number">1.3</span> Built-in pedigree structures</h2>
<p>Rather than using the <code>ped()</code> function directly, it is usually quicker and safer to build pedigrees step by step, applying the arsenal of utility functions offered by <code>pedtools</code>. A typical workflow is as follows:</p>
<ol style="list-style-type: decimal">
<li>Choose one of the basic pedigree structures as starting point.</li>
<li>Add/remove individuals as needed.</li>
<li>Modify attributes like genders and labels.</li>
</ol>
<p>You will find several examples below, but first let us list the available tools for each of the 3 steps.</p>
<p><strong>Basic pedigrees</strong><br />
The following pedigree structures serve as starting points for pedigree constructions. For parameters and details, see <code>?ped_basic</code>.</p>
<ul>
<li><code>singleton()</code>, a pedigree consisting of a single individual</li>
<li><code>nuclearPed()</code>, a nuclear pedigree (parents+children)</li>
<li><code>halfSibPed()</code>, two sibships with one parent in common</li>
<li><code>linearPed()</code>, a straight line of successors</li>
<li><code>cousinPed()</code>, cousins of specified degree/removal</li>
<li><code>halfCousinPed()</code>, half cousins of specified degree/removal</li>
<li><code>ancestralPed()</code>, a family tree containing the ancestors of a single person</li>
<li><code>selfingPed()</code>, a series of consecutive self-matings</li>
</ul>
<p>There are also more specialized structures, including double cousins and breeding schemes like consecutive matings between full siblings. Look them up in <code>?ped_complex</code> if you are interested.</p>
<p><strong>Add/remove/extract individuals</strong><br />
The functions below are used to modify an existing <code>ped</code> object by adding/removing individuals, or extracting a sub-pedigree. For details, see <code>?ped_modify</code>.</p>
<ul>
<li><code>addChildren()</code>, with special cases <code>addSon()</code> and <code>addDaughter()</code></li>
<li><code>addParents()</code></li>
<li><code>removeIndividuals()</code></li>
<li><code>branch()</code></li>
<li><code>subset()</code></li>
</ul>
<p><strong>Edit labels and attributes</strong><br />
The following functions modify various attributes of a <code>ped</code> object. See <code>?ped_modify</code> for parameters and details.</p>
<ul>
<li><code>setSex()</code></li>
<li><code>swapSex()</code></li>
<li><code>relabel()</code></li>
<li><code>setFamid()</code></li>
</ul>
</div>
<div id="examples-of-pedigree-construction" class="section level2" number="1.4">
<h2 number="1.4"><span class="header-section-number">1.4</span> Examples of pedigree construction</h2>
<div id="example-1-trio" class="section level3 unnumbered" number>
<h3 class="unnumbered" number>Example 1: Trio</h3>
<p>As our first example we will recreate the <code>trio</code> pedigree without using the <code>ped()</code> constructor. To give a hint of the flexibility, we show 3 alternative ways to code this.</p>
<p><strong>Alternative A</strong><br />
The obvious starting point is <code>nuclearPed()</code>, with <code>nch = 1</code> to indicate 1 child. By default, this creates a trio with numeric labels (father=1; mother=2; child=3) and a male child. Hence we fix the gender with <code>swapSex()</code>, and edit the labels with <code>relabel()</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>trio2 =<span class="st"> </span><span class="kw">nuclearPed</span>(<span class="dt">nch =</span> <span class="dv">1</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>trio2 =<span class="st"> </span><span class="kw">swapSex</span>(trio2, <span class="dt">ids =</span> <span class="dv">3</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a>trio2 =<span class="st"> </span><span class="kw">relabel</span>(trio2, <span class="dt">new =</span> <span class="kw">c</span>(<span class="st">&quot;fa&quot;</span>, <span class="st">&quot;mo&quot;</span>, <span class="st">&quot;girl&quot;</span>))</span></code></pre></div>
<p><strong>Alternative B (quickest and best)</strong><br />
The previous approach can be condensed into a one-liner, since <code>nuclearPed()</code> allows an alternative syntax in which child genders and labels are specified directly:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>trio3 =<span class="st"> </span><span class="kw">nuclearPed</span>(<span class="dt">father =</span> <span class="st">&quot;fa&quot;</span>, <span class="dt">mother =</span> <span class="st">&quot;mo&quot;</span>, <span class="dt">children =</span> <span class="st">&quot;girl&quot;</span>, <span class="dt">sex =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><strong>Alternative C</strong><br />
Here is another possibility. We start by creating the father as a singleton, and then add the daughter:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>trio4 =<span class="st"> </span><span class="kw">singleton</span>(<span class="st">&quot;fa&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a>trio4 =<span class="st"> </span><span class="kw">addDaughter</span>(trio4, <span class="dt">parent =</span> <span class="st">&quot;fa&quot;</span>, <span class="dt">id =</span> <span class="st">&quot;girl&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co">#&gt; Mother: Creating new individual with ID = NN_1</span></span>
<span id="cb11-4"><a href="#cb11-4"></a>trio4 =<span class="st"> </span><span class="kw">relabel</span>(trio4, <span class="dt">old =</span> <span class="st">&quot;NN_1&quot;</span>, <span class="dt">new =</span> <span class="st">&quot;mo&quot;</span>)</span></code></pre></div>
<p>Note that <code>addDaughter()</code> automatically created the mother as “NN_1”, so we needed to relabel her.</p>
</div>
<div id="example-2-an-inbred-child" class="section level3 unnumbered" number>
<h3 class="unnumbered" number>Example 2: An inbred child</h3>
<p>This time we will create this inbred family:</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAS1BMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6kNtmAABmtv+QOgCQZgCQ2/+2ZgC2/7a2///bkDrb////tmb/25D//7b//9v////qHGfVAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAF/ElEQVR4nO2d7XoaNxQGN0nd1tQtLQXM/V9pwSR50uDd0dcRWjPzyzHSeaVBu+yCg6aTLDLdewCjoyBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgYFnQ9I22mYvERJVXWCz+w8/FCVmZTZd0kwnM9vu5YsuBFz+YldJmAnPdbn/fbhF1EdRqAjOd3v113NjbhzSbQI6g0MF3ySgonlVoPYLaTSBvKYaOvmFEwwkoqKDU0rVRdkBukRYJLSeQKSh6/PEBufUVVFBLQdBeQdB+sX4LFgcUHHD/FVTZIjpAQdBCQdAiWFBC+bEFRV8oKmi5Skr5AQSNfbP6CIKq3k4ZQVD0+0G5CyuzVQdBNRNI65P78p/Tqoegigkk9rm5Xk/9UGAQQeUTSE6bFv5VVjGlRTNBxRNIT/uunG4wMwfSS1DhBHLSEu6+Myv+ULXgVj8x4CYqo0dN2r0q9g4ITlNQ94q9A4LTFNS54F0i4sK6DL6noVUK6mmocVSvkTf+U8alpIGrDRHVNGfVZ8+imNeX83X6c5taMxx++6ekW2rcbpo+/VUYwCmvL+fiu+mXFrVmOG4+lwpKydudq++rDC2GHJ7++BpSXWqG/TSVC+LE4+b58jSnPsNFGafUZ6DMz/O+QlBaZrigbdIUSs+ZVYKSUndxh9gb+6SzdPFrSp0gzt1nvMoUBiQt0HsJSrhkfH358m9FfXg8bf1UXJTUCkp5jmuOMai+i/bTQBCmX1+MQ4rvpqTSNRe1DQTN51/VxL1QHp6i108bQfMj2J5PP68vNWfpxcntrp8J0CF8f0ELhs7jrzjAmtzy9X0LdIawQdQXHsJP3DCq6w7iJ+w9tNqqw/g5BY3lIwkKGUxlzaH8hAynruRgfsb7HFRBob0DUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQ8nKAmX3XS7T9fte49QICCqusrKHQECgrtPUCAgqrrKyh0BAoK7T1AgIKq6ysodAQKCu09QICCqusrKHQECgrtPUCAgqrrKyh0BAoK7T1AgIKq6ysodAQKum1f9SlcSVZI0fQJZAsqfCyb78Nt7Ch3AqMK+l+tloo+hqDbb/JtVzrzsSEFvVOo2SL6CILeraMgKhNbffax9QhqVH79gmarKIiKRNdXEBVREBRZu6AmAWsXtFhj9YKi/y539YKqeie0UNBHFxR8jD3Aq1hCmoLKmygI2ihouc0j3KwmpT3y+0Fpae+3UtBiqwd5Tzox7bZZKz0fRNDP6+VxPhdLT5tmfq4mXFDpnXhuWk7ZHHIn0DYdR9c1rQkKAhQEKAhQEKAgQEGAggAFAQoC6oa8zdx3KSvtuLlc/ddsngYcnhLKVwnaT5GCDr9W7d6I7M+DP27IUI2g81McKajNpjazXLf8wQ0cawTtvvwZKWgXeHSdkhdohaBzQug5aPv7VLn93iL7z39vEuqXC7os0UhBx82l+DbM0O6ygSnvoFkuaHcef+gKeiPuRHTdPBPrFwt6O4TjBVVtv7fI9fSM9YsFfd2cLW93swJBYa/117WD9ce9UGywP+Eix01S/XEFnbaX82fcSfrtJMr7Hw4sqH5/QmCfchnhzSqgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgICB6yKXfXz8M4YLuGd4CBQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAvxsHljDGO+KggAFAQoCFAQoCFAQoCBAQYCCgG6C9iWbJQ1A3xXEey4OR19BuOXieHQVdHiK2+spiq6CtutbQF0FHTfrW0BdBe0/xW69G0JPQbnbtQ1BR0GrPMJ6CorbETSSjoJWeQrqKWiFV4knb1YRBQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCgP8AEBrQu6KsajUAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p><strong>Alternative A</strong><br />
One approach is to first create individuals 1-6 as half sibships, with 1 child on the left side, and 2 children on the right. After this, we use <code>addChildren()</code> to add the inbred child.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>x1 =<span class="st"> </span><span class="kw">halfSibPed</span>(<span class="dt">nch1 =</span> <span class="dv">1</span>, <span class="dt">nch2 =</span> <span class="dv">2</span>, <span class="dt">sex1 =</span> <span class="dv">1</span>, <span class="dt">sex2 =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">1</span>)</span>
<span id="cb12-2"><a href="#cb12-2"></a>x1 =<span class="st"> </span><span class="kw">addChildren</span>(x1, <span class="dt">father =</span> <span class="dv">4</span>, <span class="dt">mother =</span> <span class="dv">5</span>, <span class="dt">nch =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><strong>Alternative B</strong><br />
We could also view the half siblings 4 and 5 as half cousins of degree 0. The <code>halfCousinPed()</code> function accepts an option <code>child = TRUE</code> adding an inbred child. The labels will be different with this approach, so you should plot the pedigree after each command to see who-is-who. Also, we must relabel in the end.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>x2 =<span class="st"> </span><span class="kw">halfCousinPed</span>(<span class="dv">0</span>, <span class="dt">child =</span> T)</span>
<span id="cb13-2"><a href="#cb13-2"></a>x2 =<span class="st"> </span><span class="kw">addChildren</span>(x2, <span class="dt">father =</span> <span class="dv">1</span>, <span class="dt">mother =</span> <span class="dv">4</span>, <span class="dt">nch =</span> <span class="dv">1</span>)</span>
<span id="cb13-3"><a href="#cb13-3"></a>x2 =<span class="st"> </span><span class="kw">relabel</span>(x2, <span class="dt">old =</span> <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>,<span class="dv">7</span>,<span class="dv">6</span>), <span class="dt">new =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">7</span>))</span></code></pre></div>
<p><strong>A note about the order of pedigree members</strong></p>
<p>Although both <code>x1</code> and <code>x2</code> reproduce exactly the plot shown above, they are not identical objects:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw">identical</span>(x1, x2)</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>The reason is the <em>order</em> in which the individuals are stored. For <code>x1</code> the ordering is the natural sequence <code>1,2,3,4,5,6,7</code>, but for <code>x2</code> our construction process has produced a slightly different order:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>x2</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co">#&gt;  id fid mid sex</span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="co">#&gt;   1   *   *   1</span></span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="co">#&gt;   2   *   *   2</span></span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">#&gt;   4   1   2   1</span></span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="co">#&gt;   3   *   *   2</span></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="co">#&gt;   5   1   3   2</span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="co">#&gt;   7   4   5   1</span></span>
<span id="cb15-9"><a href="#cb15-9"></a><span class="co">#&gt;   6   1   3   1</span></span></code></pre></div>
<p>The internal ordering is usually of little importance in applications.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> However, if you get annoyed by “wrong” orderings such as for <code>x2</code> above, you can use <code>reorderPed()</code> to permute the pedigree any way you like. In fact, the default action of this function is to permute into the natural order of the labels, which is exactly what we need to make <code>x2</code> identical to <code>x1</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>x2 =<span class="st"> </span><span class="kw">reorderPed</span>(x2)</span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="kw">identical</span>(x1, x2)</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div id="example-3-a-complex-family-tree" class="section level3" number="1.4.1">
<h3 number="1.4.1"><span class="header-section-number">1.4.1</span> Example 3: A complex family tree</h3>
<p>For our final example we consider a complicated family tree extending both upwards and downwards from a single person.</p>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAVAAAAFjCAMAAABc2KPoAAAATlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6kNtmAABmADpmtv+QOgCQZgCQ2/+2ZgC2/7a2///bkDrb////tmb/25D//7b//9v////XrGT5AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALQklEQVR4nO2d6YLbNgyE1aRpu27duFW1a7//i9ayFK+9qwMgBwTIzPxKInkw+EwdvpDuQkHVeQdoTQQKFoGCRaBgEShYBAoWgYJFoGARKFgEChaBgkWgYBEoWAQKFoGCRaBgEShYBArWEtBuS6i6WL/NyOga23sJ/02yTazHUJhuN10QJZ4iawEZA/2QB7KArIF2m3/d22QL9LMDvmG0/yeH9VVQHOjySdvCVbYx1X3NNARQI1eUuypyaaArjzeyhZjrIhcGqjr1IHzzvZWRCTTZOgJQ5e0GxNjOWnxuJVDho6XwCFT4YHegPm0TaDXOEKBZSk2+q21nq8wFVqj+rS2pc0bdjFDuhzyBFgUa8ipPoOpHE6hedkCdXymB04lFoNoHpvrue4MjFwa6tkMmT0ugysilgSqeV40sgeoihwDq+pmSoLYGEhpoCtFsnDlARU/mp30Kfuq5982KT2E8P5ffD7tsobz9ywJ6EcTsVv6crhSgYprTzst/FhXLBXrZDTtv3nsDSi41UH3l98zqJMnvwC0F2NgMgvnDTZ45tbQkM7CpxQCW9mkyDmXdcTCm9nEKtBuGaZEgZXoNwLRUhGKNujItWLxkl05My5Yt3GJxpuULli13KXv4ORwS7R6EXicYj6IX+9trt0vgbt3X3/+1qHs+Xnt+sXDusS9qn3X6+t/OHnul3w5fLICej798vy6jX/HO/TVvdzU30dDlAh26zgTo67e/rrV7vPfb4brsu6PBM3UzzwU6dC+DCdCp9mC0kqyA9l//zj7kL5ZAT0beRof862/f88+hpkBtrkqDke/5+AK4KBkCHSyuSTd1x92+E9RfTSMDHexubkxOztcDHnHbZAa0717sgN5uIsDq589Tdpy9gPbXYBZAp/sxs8Mq7Ap9/fZi9LJ3bLk7mlyVLoGBzsePxe3Naf+wzDDPB2oou3OolfHPWrvJpjxrN9mUZ+0mm/Ks3WRTnrWbbMqzdpNNedZusinP2k025Vm7yaY8azfZlGftJpvyrN1kU561m2zKs3aTTXnWbrKpxWKYH+yIqqDc1NXDFAP9pO4HSi+mbQF9huiCtCmgn38yl++ZncGtWH6SBYfGvxJuC3TRoDjRdoCuPL40UQIFqxmgqw9v+XdKhkA3Hl2WKIGC1QjQrQcTKNiZQMHOBAp2/mmBZim5LFpxim0n2cnJFarcSKD6YgQKLpYFlDf2yo0Eqi+WB5Rvjug2CnKu7UGgCVs39uD7oSlb13fhRyBJW1f3Kf9JcktA+TGyfKMwJ7/oIN0oznnf7+f4Kk76l8XEOdFfPdPKrbBSteSsJmgtOasJWkvOaoLWkrOaoLXkrCZoLTmrCVpLzmqC1pKzmqC15KwmaC053YK+HcbXh/LxjPKcr99MRhFL5QX0NutQIXHO4et/12fLj6gXUO0EOGnO823sncFwZ6m8gPbKNSTNqV35cHkBPf3RqSb7SnMOX/45GI0MlskJ6NthHHN4kjcuzdmPY+LPRlPCJXK9HVGcSMVAbxM0DSdx78kVqGKyrxjoDaXFyGChfIHKryDyc+i/OmO0nIBOS8jgkH87KI3RcrvKj5cNg4vSbdr82WwG877cDnnlqGR5zkF3P4ZWLW861JKzmqC15KwmaC05qwlaS85qgtaSs5qgteSsJmgtOasJWkvOaoLWkrOaoLXkrCZoLTmrCVpLzmqC1pKzmqC15KwmaC05qwlaS85qgtaSExkUPJ0640dN+hoYu5slyuchFMYz42d3wgKPkWFIQc13W39N9EzemOQPW/Q2JoB4xkAXMmd7olwWPbKNbYEuGRgtLpRHrrMp0JXI+DMJ0CIy0LWHhwC66mDUNMB7/dHGp/48BwJFG1h1bWhNoGhn25uTTAMCRRuYtU2gYOOfFmiWNqtaOQcHmlM7Y4Vm1CVQcN3QQPe8CVRpQKBgg6hAY79SyngtT6A6BwLFWuw7uwGN/X7omkdkoGuRIwMVGDsCXdwFwcLKRPRcewJdyix4VIpvkku39de02sZAP+4U6nP5Dz5ST1+gTwijfXPk8vT+D6Z2+ltR8mdUHVngibOKU7vJpjxrN9mUZ+0mm/Ks3WRTnrWbbMqzdpNNedZusinP2k025Vm7yaY8azfZlGftVprqO9WwMGntYX69LbeWNvX6+zgi73zsgLPIgEDH0Z2Dzfi1i25OtdD47TDOHDwfr7l72GBxHNBprOwJ3/dNqjHVMuPryh9Np0GRsDHYlQB9/aY5JkXGQ/fyMBVz+AU0cbSSQ/6kWkBS4wegugKA2hJdjyHV8yyv/XZQXTT0QAfYVQkIdHySVYemvLbygFQDHXD/2QUOqP7kLq99Goe0y6UFilufSKDTKtKsJXFt5RGvBdojJ+KiV6hmnLS4tnaKug5or3o1gqotkN05VHtPowKquyOD1ZbopHwJJ66tvetWAe2nF7YB70Pj1G6yKc/aTTblWbvJpjxrN9mUZ+0mm/Ks3WRTnrWbbMqzdpNNedZusinP2k025Vm7yaY8azfZlGftJpvyrN1kU561m2zKs3a0ptJ/IKSoi/19UMaPmrQ19vYS/ptkm1z3WKhuN12wJRLmFJkDfTKpYmDzh5Qbjg5APxFEeCZvTPNfXwXlgS445C9SW6BL+dZMiwNdNIgNVBW5NNCVxxvZQsx1kYMANWra1Hv53wsDVZ3LEb753srIZYEqbzcgxnbW4nMrgcoeTaBoZyk8AhU+2B2oT9sEWo0zBGiWUpPvatvZKnOBFbq9OeQKzejIHWjIixKBqh9NoHrZAXW+sSfQskAjvjkCjlwY6OoOcYEqIwcBmsnTA6jiny2Byk88KlkC1UUuDnRpF9cP6SSRlzIr7GyBRvsYWVT9U2TVpcoY6Ie9vL/oIIz8nFlnZw401FdxpAEeIqtfPaW+YaQ4ekVucqW/yYWODGvJxK6EYhMg0Nh2JRSbAIHGtiuh2AQINLZdCcUmQKCx7UooNgEC3bSbZhRrxljK0/W42VOPmiMr3KWRZ+fLzmjDLbtpRvFtUJp0WJoY6OlqqB1pJ9AceRxFBo48O1/G9ZUIdJ5RfD6O49eEo4ClQKd5i71uiuW+5sg392mE8b5kkWfny2ieCPTHjGLVXFApUP1sTJHr41hlKNAH5/7r36mH/Dy98LfvF/EYSyVQ6KDJyfieswefpWbnK47kc+jkoVpM4jP8tOztgA7iQZE6oOMJMCLQ+aJkuULPR9kZWgd0PO1nAjU55G/DRr/88yf8xkk/ml4F9AYjF6jFRWnS+60dTA9AhXdlKqDz9NaNI2sfqMlt0yT4bdPTGkBfR+92mSvU5Mb+9n+loO+aRk2Rx5anhbCv8kAtXseNt8cmLz3nyKfNw/JJJYHqxTdHYtuVUGwCBBrbroRiEyDQ2HYlFJsAgca2K6HYBAg0tl0JxSZAoLHtSig2AQKNbVdCsQkQaGy7EopNgECz7TJ+1OQl08imBlGBJm7LNs82IFCwAYGCDQgUbECgYAMCBRsQKNiAQMEGBAo2IFCwAYGCDQgUbECgYAMCBRsQKNiAQMEGBAo2IFCwAYGCDQgUbECgYAMCBRsQKNiAQMEGBAo2IFCwAYGmGPC7Tc/miITUuwgULAIFi0DBIlCwCBQsAgWLQMEiULAIFCwCBQsH9D4Z0GBEoJGmpOOQs66TzfbbFwzofejz+/Tn6JqTTvM8UUIBvQ99fp/+HF0/kgqnOAoFAnof+vw0Vzq07kl71NF+E+4ceudYCdDLfTbmH514YPO+CPTtME6vPKGIEuinP+aJQCfB/hsNAp0Eu3ciUNUE7H0R6DQAnRclhPQTsPfFN0fAIlCwCBQsAgWLQMEiULAIFCwCBYtAwSJQsP4HrhJnOuFtbl8AAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>We will use this example to demonstrate the <code>mergePed()</code> function. When this function is given two pedigrees, it “glues together” members with matching ID labels, and checks that the result is a valid pedigree.</p>
<p>The hardest part of using <code>mergePed()</code> is to get the labelling right; this will almost always involve the <code>relabel()</code> function. To keep track of the labels, you should plot after each new line of code. Here is how the pedigree was created:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># Top part</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>x =<span class="st"> </span><span class="kw">ancestralPed</span>(<span class="dt">g =</span> <span class="dv">2</span>) <span class="co"># bottom person is `7`</span></span>
<span id="cb17-3"><a href="#cb17-3"></a></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co"># Bottom part</span></span>
<span id="cb17-5"><a href="#cb17-5"></a>y =<span class="st"> </span><span class="kw">halfCousinPed</span>(<span class="dt">degree =</span> <span class="dv">1</span>) </span>
<span id="cb17-6"><a href="#cb17-6"></a>y =<span class="st"> </span><span class="kw">swapSex</span>(y, <span class="dv">4</span>)</span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; Changing sex of spouses as well: 3</span></span>
<span id="cb17-8"><a href="#cb17-8"></a>y =<span class="st"> </span><span class="kw">relabel</span>(y, <span class="dt">new =</span> <span class="dv">7</span><span class="op">:</span><span class="dv">15</span>) <span class="co"># top person becomes `7`</span></span>
<span id="cb17-9"><a href="#cb17-9"></a></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co"># Merge</span></span>
<span id="cb17-11"><a href="#cb17-11"></a>z =<span class="st"> </span><span class="kw">mergePed</span>(x, y)</span></code></pre></div>
</div>
</div>
<div id="pedigree-subsets" class="section level2" number="1.5">
<h2 number="1.5"><span class="header-section-number">1.5</span> Pedigree subsets</h2>
<p>Pedtools offers a range of utility functions for identifying subsets of pedigree members. These come in two flavours: 1) members with certain <em>global</em> property, and 2) members with a certain relationship to a given individual.</p>
<p><strong>Pedigree members with a certain property</strong><br />
Each of the following functions returns a vector specifying the members with the given property.</p>
<ul>
<li><code>founders()</code></li>
<li><code>nonfounders()</code></li>
<li><code>leaves()</code></li>
<li><code>males()</code></li>
<li><code>females()</code></li>
<li><code>typedMembers()</code></li>
<li><code>untypedMembers()</code></li>
</ul>
<p>By default, the output of these functions is a character vector containing ID labels. However, adding the option <code>internal = TRUE</code> will give you an integer vector instead, reporting the internal indices of the members. This is frequently used in the source code of <code>pedtools</code>, but is usually not intended for end users of the package.</p>
<p><strong>Relatives of a given individual</strong><br />
The functions below take as input a <code>ped</code> object and the label of a single member. They return a vector of all members with the given relation to that individual.</p>
<ul>
<li><code>father()</code></li>
<li><code>mother()</code></li>
<li><code>parents()</code></li>
<li><code>grandparents()</code></li>
<li><code>children()</code></li>
<li><code>spouses()</code></li>
<li><code>siblings()</code></li>
<li><code>cousins()</code></li>
<li><code>nephews_nieces()</code></li>
<li><code>ancestors()</code></li>
<li><code>descendants()</code></li>
<li><code>unrelated()</code></li>
</ul>
</div>
</div>
<div id="markers" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Markers</h1>
<p>The other main theme of the <code>pedtools</code> package (pedigrees being the first) are marker genotypes.</p>
<div id="creating-marker-objects" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> Creating marker objects</h2>
<p>Marker objects created with the <code>marker()</code> function. For example, the following command makes an empty marker associated with the <code>trio</code> pedigree:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">marker</span>(trio)</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="co">#&gt;       &lt;NA&gt;</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="co">#&gt;    fa  -/-</span></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">#&gt;    mo  -/-</span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">#&gt;  girl  -/-</span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt; ----------- </span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">#&gt; Chrom NA: NA (Mb), NA (cM)</span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">#&gt; Mutation model: No </span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">#&gt; Allele frequencies:</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">#&gt;    1   2</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">#&gt;  0.5 0.5</span></span></code></pre></div>
<p>As shown in the output, the marker is indeed empty: All pedigree members have missing genotypes, and there is no assigned name or chromosome/position. Furthermore, the last lines show that there is only one allele (named “1”), with frequency 1. For a more interesting example, let us make a SNP named “snp1”, with alleles “A” and “B”. The father is homozygous “A/A”, while the mother is heterozygous. We store it in a variable <code>m1</code> for later use.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>m1 =<span class="st"> </span><span class="kw">marker</span>(trio, <span class="dt">fa =</span> <span class="st">&quot;A&quot;</span>, <span class="dt">mo =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>), <span class="dt">name =</span> <span class="st">&quot;snp1&quot;</span>)</span></code></pre></div>
<p>This illustrates several points. Firstly, individual genotypes are specified using the ID labels. For homozygous genotypes it suffices to write the allele once. Furthermore, the different alleles occurring in the genotypes is interpreted as the complete set of alleles for the marker. Finally, these are assigned equal frequencies. Of course, this behaviour can be overridden, by declaring alleles and frequencies explicitly:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">marker</span>(trio, <span class="dt">fa =</span> <span class="st">&quot;A&quot;</span>, <span class="dt">mo =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>), <span class="dt">alleles =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>,<span class="st">&quot;C&quot;</span>), <span class="dt">afreq =</span> <span class="kw">c</span>(.<span class="dv">2</span>,.<span class="dv">3</span>,.<span class="dv">5</span>))</span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="co">#&gt;       &lt;NA&gt;</span></span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="co">#&gt;    fa  A/A</span></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co">#&gt;    mo  A/B</span></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="co">#&gt;  girl  -/-</span></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="co">#&gt; ----------- </span></span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="co">#&gt; Chrom NA: NA (Mb), NA (cM)</span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="co">#&gt; Mutation model: No </span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="co">#&gt; Allele frequencies:</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co">#&gt;    A   B   C</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="co">#&gt;  0.2 0.3 0.5</span></span></code></pre></div>
<p>The markers chromosome can be declared using the <code>chrom</code> argument, and similarly its position by <code>posMb</code> (megabases) and/or <code>posCm</code> (centiMorgan). Markers with unknown chromosome are treated as autosomal. To define an X-linked marker, put <code>chrom=23</code>. the fact that males are hemizygous on X (i.e. they have only one allele) is reflected in the printout of such markers:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>m2 =<span class="st"> </span><span class="kw">marker</span>(trio, <span class="dt">fa =</span> <span class="st">&quot;A&quot;</span>, <span class="dt">mo =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>), <span class="dt">chrom =</span> <span class="dv">23</span>, <span class="dt">name =</span> <span class="st">&quot;snpX&quot;</span>)</span></code></pre></div>
<p>A side note: It may come as a surprise that you don’t need quotes around the ID labels (which are characters!) in the above commands. This is because <code>marker()</code> uses <em>non-standard evaluation (NSE)</em>, a peculiarity of the R language which often leads to less typing and more readable code.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> Unfortunately, this doesn’t work with numerical ID labels. Thus to assign a genotype to someone labelled “1” you need quotes, as in <code>marker(trio, &quot;1&quot;=&quot;A&quot;)</code>.</p>
</div>
<div id="plotting-pedigrees-with-marker-data" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> Plotting pedigrees with marker data</h2>
<p>Including marker data in a pedigree plot is straightforward:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw">plot</span>(trio, <span class="dt">marker =</span> m1)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAXVBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6Ojo6OpA6kNtmAABmADpmZjpmtrZmtv+QOgCQOjqQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///87oUNWAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAEEUlEQVR4nO2cjXLTMBCElZAACeBSC6hx0rz/YyJZ8W8kxzr7fFbZndKphub2PklnS+5Y6pa4lHQCcwUAaQFAWgCQFgCkBQBpAUBaAJAWAKT1IQFUUCvnNsXWCxDzy3xq3cYQNgvQTzo8DFsFeLBKDMDjFDDfJoDXyO++SYCIzgYAi1IHiPIHAIMYAVYSH0DoP8KiDBoAADDvI/MXblsFELmR/Z8AEQvPlABkltPEm/d2NjTU1cfwc2J7YvLyabCpjzLYBEAXIfqxSvQqKy7+9A9PsOVeIbOvwAEgHB8A0vEBIB0fANLxPzJAodSJM/5CChpczwukLwuQscZfSiGDy1Gp/dtNm7XgLBAawOX442xmsE3CupcqPJ3HR0AfTC3sclIST+KP6nI0poXtwcL8Kw3F9XyINLAA12+5DTZnCKgAp+Zb9v5ie78M9OPTGijnzSEqQNZ+cz0Y6sdxAHMp3f8WGYEHgNBFZRSgjUOW7AiUpoKqEqJrAYAZNWBDXM+zbsgLAMy4Ctka2OV6znV0CQDafWAhYTUqbQAAaQMASBsAgGig6xvH5evfbnOp+E/VTUBXz9ijlhKXz9/tMsioPPSa0SIC9BLQn2wnBjZWfoOiWUTrrNeMFhGgl4ADCDxl8Bq8vxzMV/XTz7zbjBcNoJ9ADeDtQ6+BXbq6EatKoG3GiwbQT+A+hfxd6DWwn3AjVpVA24wXDaCfgCtiVWFMMqh20zdtq0hnvWa8SACDBNwIlMo7BD6D4v7HwcyVQNskiAQwSMABBLrQY3AvH7sHsjOw0ySIAjBIoAYopgLUu09TRXYGdpqEZCgAgwSaEfAWgceg/kUzE9sZWE/MWFEABgk0NTDxMto+v9C7L3m3SSljAsAggfoqFPtUYiFhNSptAABpAwBIGwBA2gAA0gYAkDYAgLQBAKQNACBtAABpAwBIG6QNQH7vI8aDL3KTOSsDW2g10uLz4Yu7UkcxhuWaRjxhvVFXtOKJmhCAP2jyAKt6sQRNBiAYc1UzjpgAkI4JAOmYAFgpZvL3gY8KkNBaKPnVqDdsUhsaT9y0dmSPgVPbE996Xc75XIXzqdOME00iTDiDryEASAsA0gKAtAAgLQBIixegeXFnoXNCPFppBAAQFBOAVmr3usvd8SCvav8nMQBdnYhyBzhX71QlBeDewdA1wCm5KVQdqGHf5XEAWXIAReoAyY+Aq4EiXYDBVShBAHsf2P/av6ULYFVSX2OPEl8NLHRQ3TPxjEBJfn03WtgPSAsA0uIGcEebMIoboCQeqDFZ3ACa+2LKDFAdq8AqZgD2EuAGYC8BbgD2EmAG4C8BPgB7PsrJlkAxclDiAuIdAf4SYAbgLwFegBVKAIs5cQFAWgCQFgCkBQBpAUBaAJDWPz5IfkH3HekiAAAAAElFTkSuQmCC" style="display: block; margin: auto;" /></p>
<p>The appearance of the genotypes can be tweaked in various ways, as documented in <code>?plot.ped</code>. Here’s an example:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="kw">plot</span>(trio, <span class="dt">marker =</span> m1, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="dt">skip.empty.genotypes =</span> T)</span></code></pre></div>
<pre><code>#&gt; The `skip.empty.genotypes` argument has been renamed to `skipEmptyGenotypes`, and will be removed in a future version</code></pre>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAADACAMAAABlApw1AAAAVFBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6kNtmAABmADpmZjpmtrZmtv+QOgCQkGaQ2/+2ZgC2tma225C2/7a2///bkDrb////tmb/25D//7b//9v///9DYrSiAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADtUlEQVR4nO2cC3PaMBCEhQuFtiQ4La4h/P//WcsPLMAy6OTzWunuJBllgm/16WGdnJHNJXEZdAViRQC0CIAWAdAiAFoEQIsAaBEALQKg9SUBjFcz1+0V20GAkA/rqXcbQ1gswG2l/d2wVIAHq8QABpw85ssEGDQadl8kQEBjE0BFqQME+RNAQYoAM0kPwPcHvySdRgACxF0Sn7gtFQCykP2fAAGJZ0oAmHRauHgvZ0MjzT7ur4PticXp092mPshgEQAuQvBjleAsKyz+6xe/YKudIatn4AQAxycAOj4B0PEJgI7/lQEKY7aa8SeS1+C8m6D6WIC9avyp5DM4bYzJjpe8ygWjQGQAp82vXTWCbSWse2n8w3m8B/J1NRdWB1ElnsQf1WlTmRa2BYvqu6wozrt1oIEFOP842GAxXSAF2F5/7D/fbOuXnnZ8OgfKuDEkBdj3P5oW9LXjOEB1K81+Q3rgAcB3UxkF6OOIhe2BsppB9RSSawKAiDlgQ5x3UQvyBAARdyE7B1aHPOY+OgWAbB2YSMxG0QYEQBsQAG1AALFB3i8decRaLAfoXPP6Cbt3MfXuyL7/tInQXTFcYoCra/7t72VkW+UzKPo0uojJqMUAV9cGwPuMwWPw+bauvu6LAkkBetcOwNOGHgObvLa95hQFkgL0ru0Q8jWhx8Be1faaUxRICtC7NpPY1BgvG9T76UtuZ5FTlEgI4Lg2PVCaoP1A0f57cH9TlEgI4Lg2AN4mHDRop4/dBTlFkWQArmsLUIQAdPvPahY5RVFVZACua9cDnkkwaNB9uBqJTlFUFRmA69rNgYDbaP8EI+/HvnAaiwCcCmRHYSoxlZiNog0IgDYgANqAAGgDAqANCIA2IADagABoAwKgDQiANiAA2iBtAPG5jxAPvcjXmqsyqIU2I7/p+ejFnamhFMNqDSOdsINRZ7TSiZoQwHDQ5AFm9VIJmgyAN+asZhoxCYCOSQB0TALMFDP5deCrAiSUCyWfjQ6GTWpDMxA3rR3ZY+DU9sSXmybXfK6i+dQp4o0mASaawecQAdAiAFoEQIsAaBEALV2A69Gdid4TMqCZeoAAXikB5Mas3leH5vUg7yb7kxhAXr8RpQXY1SeqkgJoTmDkHcA2uSFUv1DDnuRpAPbJARSpAyTfA80cKNIFuLsLJQhg14HsIzumC2BVyg+yB0hvDkz0orpn0umBMuIAb6C4H0CLAGgRAC0CoEUAtAiAFgHQIgBaBECLAGgRAC0CoEUAtAiAFgHQ+gcIadt1Csvt6wAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
</div>
<div id="markers-attached-to-ped-objects" class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> Markers attached to <code>ped</code> objects</h2>
<p>Although a <code>ped</code> object is needed in the creation of a <code>marker</code>, the two are independent of each other once the marker is created. In many applications it is useful to <em>attach</em> markers to their <code>ped</code> object. In particular for bigger projects with many markers, this makes it easier to manipulate the dataset as a unit.</p>
<p>To attach a marker <code>m</code> (which could be a list of several markers) to a pedigree <code>x</code>, there are two options:</p>
<ul>
<li><code>setMarkers(x, m)</code></li>
<li><code>addMarkers(x, m)</code></li>
</ul>
<p>The difference between these is that <code>setMarkers()</code> replaces all existing markers, while <code>addMarkers()</code> appends <code>m</code> to the existing ones. In our <code>trio</code> example the two are equivalent since there are no existing markers.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>trio =<span class="st"> </span><span class="kw">setMarkers</span>(trio, <span class="kw">list</span>(m1, m2))</span>
<span id="cb25-2"><a href="#cb25-2"></a>trio</span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="co">#&gt;    id fid mid sex snp1 snpX</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">#&gt;    fa   *   *   1  A/A    A</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="co">#&gt;    mo   *   *   2  A/B  A/B</span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="co">#&gt;  girl  fa  mo   2  -/-  -/-</span></span></code></pre></div>
<p><strong>Selecting and removing attached markers</strong><br />
Four closely related functions functions are useful for manipulating markers attached to a pedigree:</p>
<ul>
<li><code>selectMarkers()</code>, returns a <code>ped</code> object where only the indicated markers are retained</li>
<li><code>removeMarkers()</code>, returns a <code>ped</code> object where the indicated markers are removed</li>
<li><code>getMarkers()</code>, returns a list of the indicated markers</li>
<li><code>whichMarkers()</code>, returns the indices of the indicated markers</li>
</ul>
<p>All of these have exactly the same arguments, described in more detail in <code>?marker_select</code>. Let us do a couple of examples here. Recall that by now, our <code>trio</code> has two attached markers; the first is called “snp1”, and the other is on the X chromosome (<code>chrom = 23</code>).</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">whichMarkers</span>(trio, <span class="dt">chrom =</span> <span class="dv">23</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="co">#&gt; [1] 2</span></span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="kw">selectMarkers</span>(trio, <span class="dt">markers =</span> <span class="st">&quot;snp1&quot;</span>)</span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="co">#&gt;    id fid mid sex snp1</span></span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="co">#&gt;    fa   *   *   1  A/A</span></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co">#&gt;    mo   *   *   2  A/B</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="co">#&gt;  girl  fa  mo   2  -/-</span></span></code></pre></div>
</div>
<div id="accessing-and-modifying-individual-markers" class="section level2" number="2.4">
<h2 number="2.4"><span class="header-section-number">2.4</span> Accessing and modifying individual markers</h2>
<p>Internally, a marker object is stored as a matrix with two columns (one for each allele) and one row for each pedigree member. The matrix is numeric (for computational convenience) while the allele labels and other meta information are added as <em>attributes</em>. The most important of these are:</p>
<ul>
<li><code>alleles</code> : The allele labels, stored as a character vector.</li>
<li><code>afreq</code> : The allele frequencies, in the same order as the alleles. An error is issued if the frequencies do not sum to 1 after rounding to 3 decimals.</li>
<li><code>name</code> : The marker name, which can be any character string not consisting solely of digits.</li>
<li><code>chrom</code> : The chromosome name. This can be given as an integer, but is always converted to character. The special values “23” and “X” are recognized as the human X chromosome, which affects the way genotypes are printed.</li>
<li><code>posMb</code> : Chromosomal position given in megabases.</li>
<li><code>posCm</code> : Chromosomal position given in centiMorgan.</li>
</ul>
<p>In addition to those listed above, there are two more attributes: <code>pedmembers</code> and <code>sex</code>. They store the ID labels and genders of the pedigree associated with the marker, and are only used to empower the printing method of marker objects.</p>
<p><strong>Marker accessor functions</strong><br />
For each marker attribute listed above, there is a corresponding function with the same name for retrieving its content. These functions take as input either a <code>marker</code> object, or a <code>ped</code> object together with the name (or index) of an attached marker. This may sound a bit confusing, but a few examples will make it clear!</p>
<p>Recall that our marker “snp1” exists in two copies: One is stored in the variable <code>m1</code>, while the other is attached to <code>trio</code>. In both cases we can extract the allele frequencies with the function <code>afreq()</code>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="kw">afreq</span>(m1)</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="co">#&gt;   A   B </span></span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="co">#&gt; 0.5 0.5</span></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="kw">afreq</span>(trio, <span class="dt">marker =</span> <span class="st">&quot;snp1&quot;</span>)</span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="co">#&gt;   A   B </span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co">#&gt; 0.5 0.5</span></span></code></pre></div>
<p>We can also <em>modify</em> the frequencies using this syntax. To avoid confusion about the allele order, the frequencies must be named with the allele labels (just as in the output of <code>afreq()</code> above).</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">afreq</span>(trio, <span class="dt">marker =</span> <span class="st">&quot;snp1&quot;</span>) =<span class="st"> </span><span class="kw">c</span>(<span class="dt">A =</span> <span class="fl">0.9</span>, <span class="dt">B =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<p>In addition to the functions getting and setting marker attributes, there is one more important marker accessor, namely <code>genotype()</code>. This returns the genotype of a specified individual, and can also be used to modify genotypes. As the others, it can be applied to marker objects directly, or to pedigrees with attached markers. Here we show a few examples of the latter type:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="kw">genotype</span>(trio, <span class="st">&quot;snpX&quot;</span>, <span class="dt">id =</span> <span class="st">&quot;girl&quot;</span>)</span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="co">#&gt; [1] NA NA</span></span>
<span id="cb29-3"><a href="#cb29-3"></a><span class="kw">genotype</span>(trio, <span class="st">&quot;snpX&quot;</span>, <span class="dt">id =</span> <span class="st">&quot;girl&quot;</span>) =<span class="st"> &quot;A&quot;</span></span>
<span id="cb29-4"><a href="#cb29-4"></a>trio</span>
<span id="cb29-5"><a href="#cb29-5"></a><span class="co">#&gt;    id fid mid sex snp1 snpX</span></span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="co">#&gt;    fa   *   *   1  A/A    A</span></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co">#&gt;    mo   *   *   2  A/B  A/B</span></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="co">#&gt;  girl  fa  mo   2  -/-  A/A</span></span></code></pre></div>
</div>
<div id="gettingsettingmodifying-many-markers-simultaneously" class="section level2" number="2.5">
<h2 number="2.5"><span class="header-section-number">2.5</span> Getting/setting/modifying many markers simultaneously</h2>
Several functions in <code>pedtools</code> are indented for modifying many (or all) markers at the same time. Their purpose and typical use cases are summarised in the table below. The argument <code>x</code> always denotes a <code>ped</code> object.
<table>
<thead>
<tr>
<th style="text-align:left;">
Use …
</th>
<th style="text-align:left;">
When you want to …
</th>
<th style="text-align:left;">
For example to …
</th>
</tr>
</thead>
<tbody>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Get</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>getAlleles(x)</code>
</td>
<td style="text-align:left;width: 8cm; ">
extract all alleles as a matrix.
</td>
<td style="text-align:left;">
do summary stats on the marker alleles
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>getFrequencyDatabase(x)</code>
</td>
<td style="text-align:left;width: 8cm; ">
extract allele frequencies as a data.frame in <em>allelic ladder</em> format.
</td>
<td style="text-align:left;">
transfer to other objects, or write the database to a file
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>getMarkers(x)</code>
</td>
<td style="text-align:left;width: 8cm; ">
extract list of marker objects. Each marker is a <code>N * 2</code> allele matrix (<code>N = pedsize(x)</code>) with locus annotations as attributes
</td>
<td style="text-align:left;">
do computations
</td>
</tr>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Set</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>setAlleles(x, ...)</code>
</td>
<td style="text-align:left;width: 8cm; ">
replace the genotypes of <code>x</code> without changing the locus attributes.
</td>
<td style="text-align:left;">
erase all genotypes
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>setFrequencyDatabase(x, db)</code>
</td>
<td style="text-align:left;width: 8cm; ">
replace all allele frequencies without changing the genotype data. The input is a data.frame in <em>allelic ladder</em> format. Conceptually equivalent to <code>setMarkers(x, alleleMatrix = getAlleles(x), locusAnnotations = db)</code>.
</td>
<td style="text-align:left;">
change the frequency database
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>setMarkers(x, ...)</code>
</td>
<td style="text-align:left;width: 8cm; ">
attach marker objects with or without genotype data. Locus attributes are indicated as a list; genotypes as a matrix or data.frame.
</td>
<td style="text-align:left;">
prepare joint manipulation of a pedigree and marker data
</td>
</tr>
<tr grouplength="2">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Convert</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>as.data.frame(x)</code>
</td>
<td style="text-align:left;width: 8cm; ">
convert <code>x</code> to a data.frame, with pedigree columns in standard format followed by genotype columns. One column per marker, with genotype format <code>a/b</code> and missing alleles indicated as <code>-</code>.
</td>
<td style="text-align:left;">
pretty-print ped objects
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>as.matrix(x)</code>
</td>
<td style="text-align:left;width: 8cm; ">
convert <code>x</code> to a <em>numerical</em> matrix, with additional info attached as attributes.
</td>
<td style="text-align:left;">
modify a pedigree with marker data
</td>
</tr>
<tr groupLength="1">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Transfer</strong>
</td>
</tr>
<tr>
<td style="text-align:left;width: 5cm; ">
<code>transferMarkers(from, to)</code>
</td>
<td style="text-align:left;width: 8cm; ">
transfer genotypes and attributes between pedigree objects (or lists of such).
</td>
<td style="text-align:left;">
transfer simulated marker data
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>There is an important exception to this: Certain algorithms in pedigree analysis work “top-down”, in the sense that parents must be treated before their children. For this reason, many implementations require, for simplicity, that the individuals are stored in this fashion, i.e. that parents always precede their children. <code>pedtools</code> offers a special reordering function to ensure this, <code>parents_before_children()</code>, which you will find utilised in the source code of packages like <code>ribd</code> and <code>ibdsim2</code>.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>You may have come across NSE before, for instance when using <code>subset()</code> on a data.frame. To learn more about NSE, I highly recommend this book chapter by Hadley Wickham:<br />
<a href="http://adv-r.had.co.nz/Computing-on-the-language.html" class="uri">http://adv-r.had.co.nz/Computing-on-the-language.html</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
